<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>NIH / NCI Genomic Data Commons Access • GenomicDataCommons</title>
<!-- jquery --><script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://maxcdn.bootstrapcdn.com/bootswatch/3.3.7/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script><!-- Font Awesome icons --><link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">
<!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="jquery.sticky-kit.min.js"></script><script src="pkgdown.js"></script><!-- mathjax --><script src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><!-- Google analytics --><script>
  (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
  (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
  m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
  })(window,document,'script','https://www.google-analytics.com/analytics.js','ga');

  ga('create', 'UA-93043521-1', 'auto');
  ga('send', 'pageview');

</script>
</head>
<body>
    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">GenomicDataCommons</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="reference/index.html">Reference</a>
</li>
<li>
  <a href="articles/index.html">Articles</a>
</li>
<li>
  <a href="news/index.html">News</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="http://github.com/Bioconductor/GenomicDataCommons">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="contents col-md-9">
    
<div id="what-is-the-gdc" class="section level1">
<div class="page-header"><h1 class="hasAnchor">
<a href="#what-is-the-gdc" class="anchor"></a>What is the GDC?</h1></div>

<p>The National Cancer Institute’s (NCI’s) Genomic Data Commons (GDC) is a data sharing platform that promotes precision medicine in oncology. It is not just a database or a tool; it is an expandable knowledge network supporting the import and standardization of genomic and clinical data from cancer research programs.</p>
<p>The GDC contains NCI-generated data from some of the largest and most comprehensive cancer genomic datasets, including The Cancer Genome Atlas (TCGA) and Therapeutically Applicable Research to Generate Effective Therapies (TARGET). For the first time, these datasets have been harmonized using a common set of bioinformatics pipelines, so that the data can be directly compared.</p>
<p>As a growing knowledge system for cancer, the GDC also enables researchers to submit data, and harmonizes these data for import into the GDC. As more researchers add clinical and genomic data to the GDC, it will become an even more powerful tool for making discoveries about the molecular basis of cancer that may lead to better care for patients.</p>
<p>The <a href="https://gdc.cancer.gov/developers/gdc-data-model/gdc-data-model-components">data model for the GDC is complex</a>, but it worth a quick overview. The data model is encoded as a so-called property graph. Nodes represent entities such as Projects, Cases, Diagnoses, Files (various kinds), and Annotations. The relationships between these entities are maintained as edges. Both nodes and edges may have Properties that supply instance details. The GDC API exposes these nodes and edges in a somewhat simplified set of <a href="https://en.wikipedia.org/wiki/Representational_state_transfer">RESTful</a> endpoints.</p>
</div>
<div id="quickstart" class="section level1">
<h1 class="hasAnchor">
<a href="#quickstart" class="anchor"></a>Quickstart</h1>
<p>This software is in development and will <em>likely change in response to user feedback</em>. To report bugs or problems, either <a href="https://github.com/Bioconductor/GenomicDataCommons/issues">submit a new issue</a> or submit a <code>bug.report(package='GenomicDataCommons')</code> from within R (which will redirect you to the new issue on GitHub).</p>
<div id="installation" class="section level2">
<h2 class="hasAnchor">
<a href="#installation" class="anchor"></a>Installation</h2>
<p>Installation is available from GitHub as of now.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">source</span>(<span class="st">'https://bioconductor.org/biocLite.R'</span>)
<span class="kw">biocLite</span>(<span class="st">'Bioconductor/GenomicDataCommons'</span>)</code></pre></div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(GenomicDataCommons)</code></pre></div>
</div>
<div id="check-basic-functionality" class="section level2">
<h2 class="hasAnchor">
<a href="#check-basic-functionality" class="anchor"></a>Check basic functionality</h2>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">GenomicDataCommons<span class="op">::</span><span class="kw"><a href="reference/status.html">status</a></span>()</code></pre></div>
</div>
<div id="find-data" class="section level2">
<h2 class="hasAnchor">
<a href="#find-data" class="anchor"></a>Find data</h2>
<p>The following code builds a <code>manifest</code> that can be used to guide the download of raw data. Here, filtering finds gene expression files quantified as raw counts using <code>HTSeq</code> from ovarian cancer patients.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(magrittr)
ge_manifest =<span class="st"> </span><span class="kw"><a href="reference/query.html">files</a></span>() <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">    </span><span class="kw"><a href="reference/filtering.html">filter</a></span>( <span class="op">~</span><span class="st"> </span>cases.project.project_id <span class="op">==</span><span class="st"> 'TCGA-OV'</span> <span class="op">&amp;</span>
<span class="st">                </span>type <span class="op">==</span><span class="st"> 'gene_expression'</span> <span class="op">&amp;</span>
<span class="st">                </span>analysis.workflow_type <span class="op">==</span><span class="st"> 'HTSeq - Counts'</span>) <span class="op">%&gt;%</span>
<span class="st">    </span><span class="kw"><a href="reference/manifest.html">manifest</a></span>()</code></pre></div>
</div>
<div id="download-data" class="section level2">
<h2 class="hasAnchor">
<a href="#download-data" class="anchor"></a>Download data</h2>
<p>This code block downloads the <code>r nrow(ge_manifest)</code> gene expression files specified in the query above. Using multiple processes to do the download very significantly speeds up the transfer in many cases. On a standard 1Gb connection using 10 processes, the following completes in about 15 seconds.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(BiocParallel)
<span class="kw">register</span>(<span class="kw">MulticoreParam</span>())
destdir =<span class="st"> </span><span class="kw">tempdir</span>()
fnames =<span class="st"> </span><span class="kw">bplapply</span>(ge_manifest<span class="op">$</span>id,gdcdata,
                  <span class="dt">destination_dir=</span>destdir,
                  <span class="dt">BPPARAM =</span> <span class="kw">MulticoreParam</span>(<span class="dt">progressbar=</span><span class="ot">TRUE</span>))</code></pre></div>
<p>If the download had included controlled-access data, the download above would have needed to include a <code>token</code>. Details are available in <a href="#authentication">the authentication section below</a>.</p>
</div>
<div id="metadata-queries" class="section level2">
<h2 class="hasAnchor">
<a href="#metadata-queries" class="anchor"></a>Metadata queries</h2>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">expands =<span class="st"> </span><span class="kw">c</span>(<span class="st">"diagnoses"</span>,<span class="st">"annotations"</span>,
             <span class="st">"demographic"</span>,<span class="st">"exposures"</span>)
clinResults =<span class="st"> </span><span class="kw"><a href="reference/query.html">cases</a></span>() <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">    </span>GenomicDataCommons<span class="op">::</span><span class="kw"><a href="reference/select.html">select</a></span>(<span class="ot">NULL</span>) <span class="op">%&gt;%</span>
<span class="st">    </span>GenomicDataCommons<span class="op">::</span><span class="kw"><a href="reference/expand.html">expand</a></span>(expands) <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">    </span><span class="kw"><a href="reference/results.html">results</a></span>(<span class="dt">size=</span><span class="dv">50</span>)
clinDF =<span class="st"> </span><span class="kw">as.data.frame</span>(clinResults)
<span class="kw">library</span>(DT)
<span class="kw">datatable</span>(clinDF, <span class="dt">extensions =</span> <span class="st">'Scroller'</span>, <span class="dt">options =</span> <span class="kw">list</span>(
  <span class="dt">deferRender =</span> <span class="ot">TRUE</span>,
  <span class="dt">scrollY =</span> <span class="dv">200</span>,
  <span class="dt">scrollX =</span> <span class="ot">TRUE</span>,
  <span class="dt">scroller =</span> <span class="ot">TRUE</span>
))</code></pre></div>
</div>
</div>
<div id="basic-design" class="section level1">
<h1 class="hasAnchor">
<a href="#basic-design" class="anchor"></a>Basic design</h1>
<p>This package design is meant to have some similarities to the “hadleyverse” approach of dplyr. Roughly, the functionality for finding and accessing files and metadata can be divided into:</p>
<ol>
<li>Simple query constructors based on GDC API endpoints.</li>
<li>A set of verbs that when applied, adjust filtering, field selection, and faceting (fields for aggregation) and result in a new query object (an endomorphism)</li>
<li>A set of verbs that take a query and return results from the GDC</li>
</ol>
<p>In addition, there are exhiliary functions for asking the GDC API for information about available and default fields, slicing BAM files, and downloading actual data files. Here is an overview of functionality[^1].</p>
<ul>
<li>Creating a query
<ul>
<li><code><a href="reference/query.html">projects()</a></code></li>
<li><code><a href="reference/query.html">cases()</a></code></li>
<li><code><a href="reference/query.html">files()</a></code></li>
<li><code><a href="reference/query.html">annotations()</a></code></li>
</ul>
</li>
<li>Manipulating a query
<ul>
<li><code><a href="reference/filtering.html">filter()</a></code></li>
<li><code><a href="reference/faceting.html">facet()</a></code></li>
<li><code><a href="reference/select.html">select()</a></code></li>
</ul>
</li>
<li>Introspection on the GDC API fields
<ul>
<li><code><a href="reference/mapping.html">mapping()</a></code></li>
<li><code><a href="reference/available_fields.html">available_fields()</a></code></li>
<li><code><a href="reference/default_fields.html">default_fields()</a></code></li>
<li><code><a href="reference/grep_fields.html">grep_fields()</a></code></li>
<li><code>field_picker()</code></li>
<li><code><a href="reference/available_values.html">available_values()</a></code></li>
<li><code><a href="reference/available_expand.html">available_expand()</a></code></li>
</ul>
</li>
<li>Executing an API call to retrieve query results
<ul>
<li><code><a href="reference/results.html">results()</a></code></li>
<li><code><a href="reference/count.html">count()</a></code></li>
<li><code><a href="reference/response.html">response()</a></code></li>
</ul>
</li>
<li>Raw data file downloads
<ul>
<li><code><a href="reference/gdcdata.html">gdcdata()</a></code></li>
<li><code><a href="reference/transfer.html">transfer()</a></code></li>
<li><code><a href="reference/gdc_client.html">gdc_client()</a></code></li>
</ul>
</li>
<li>Summarizing and aggregating field values (faceting)
<ul>
<li><code><a href="reference/aggregations.html">aggregations()</a></code></li>
</ul>
</li>
<li>Authentication
<ul>
<li><code><a href="reference/gdc_token.html">gdc_token()</a></code></li>
</ul>
</li>
<li>BAM file slicing
<ul>
<li><code><a href="reference/slicing.html">slicing()</a></code></li>
</ul>
</li>
</ul>
</div>

  </div>

  <div class="col-md-3" id="sidebar">
    <h2>Links</h2>
<ul class="list-unstyled">
<li>Browse source code at <br><a href="http://github.com/Bioconductor/GenomicDataCommons">http://​github.com/​Bioconductor/​GenomicDataCommons</a>
</li>
<li>Report a bug at <br><a href="https://github.com/Bioconductor/GenomicDataCommons/issues/new">
        https://​github.com/​Bioconductor/​GenomicDataCommons/​issues/​new</a>
</li>
</ul>
<h2>License</h2>
<p><a href="https://www.r-project.org/Licenses/Artistic-2.0">Artistic-2.0</a></p>
<h2>Developers</h2>
<ul class="list-unstyled">
<li>Martin Morgan <br><small class="roles"> Author </small> </li>
<li>Davis Sean <br><small class="roles"> Author, maintainer </small> </li>
</ul>
<div>
<h2>Dev status</h2>
<ul class="list-unstyled">
<li><a href="https://gdc.nci.nih.gov/about-gdc">Genomic Data Commons (GDC) website</a></li>
</ul>
</div>
</div>
</div>


      <footer><div class="copyright">
  <p>Developed by Martin Morgan, Davis Sean.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://hadley.github.io/pkgdown/">pkgdown</a>.</p>
</div>

      </footer>
</div>

  </body>
</html>
