---
title: "Introduction to Accessing the NCI Genomic Data Commons in Bioconductor"
author: "Sean Davis"
date: "`r Sys.Date()`"
output: 
  BiocStyle::html_document
vignette: >
  %\VignetteIndexEntry{Introduction to Accessing the NCI Genomic Data Commons in Bioconductor}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r init, results='hide', echo=FALSE, warning=FALSE, message=FALSE}
library(knitr)
```

```{r libraries, message=FALSE}
library(GenomicDataCommons)
library(DT)
```

# Background

From the [Genomic Data Commons (GDC) website](https://gdc.nci.nih.gov/about-gdc):

> The National Cancer Institute's (NCI's) Genomic Data Commons (GDC)
> is a data sharing platform that promotes precision medicine in
> oncology. It is not just a database or a tool; it is an expandable
> knowledge network supporting the import and standardization of
> genomic and clinical data from cancer research programs.
>
> The GDC contains NCI-generated data from some of the largest and
> most comprehensive cancer genomic datasets, including The Cancer
> Genome Atlas (TCGA) and Therapeutically Applicable Research to
> Generate Effective Therapies (TARGET). For the first time, these
> datasets have been harmonized using a common set of bioinformatics
> pipelines, so that the data can be directly compared.
>
> As a growing knowledge system for cancer, the GDC also enables
> researchers to submit data, and harmonizes these data for import
> into the GDC. As more researchers add clinical and genomic data to
> the GDC, it will become an even more powerful tool for making
> discoveries about the molecular basis of cancer that may lead to
> better care for patients.

# Data Model

The [data model for the GDC is complex](https://gdc.cancer.gov/developers/gdc-data-model/gdc-data-model-components), but it worth a quick overview. The data model is
encoded as a so-called property graph. Nodes represent entities such 
as Projects, Cases, Diagnoses, Files (various kinds), and Annotations.
The relationships between these entities are maintained as edges.  Both
nodes and edges may have Properties that supply instance details.  The GDC
API exposes these nodes and edges in a somewhat simplified set of [RESTful](https://en.wikipedia.org/wiki/Representational_state_transfer) endpoints.

There are several available endpoints:

- Projects
- Files
- Cases
- Annotations
- Slicing - Specific to dealing with BAM files
- _mappings - For obtaining metadata about fields associated with the various endpoints

The following sections address each in turn, including examples of usage.

# Using the GDC API

## The basic API endpoints

The Projects, Files, Cases, and Annotations endpoints are are all accessed
using functions of the same name (all lowercase). The default behavior of
each is to query the GDC and return the first 10 records with a default
set of fields. In this section, the results are simply converted to a tabular
form, but the data are not always tabular and may have one-to-many relationships, so this represents a simplification. Accessing the data more directly will be covered in later sections. 

### Projects

To get a sense of what is available in the default Project entities, we simply call the `projects()` function without any arguments. By default, the first 10 projects are returned. The returned value for projects is a `data.frame` since the Project entity is tabular. 

```{r project10}
a = projects()
datatable(a)
```

### Files

```{r files10}
fields <- c("data_type", "updated_datetime", "created_datetime",
    "file_name", "md5sum", "data_format", "acl", "access", "state",
    "data_category", "file_size", "submitter_id", "type")
a = files(fields=fields)
datatable(do.call(rbind, a))
```

### Cases

```{r cases10}
a = cases()
a             # 10 cases by default
lengths(a)    # cases have different elements
a[1]          # summary
head(a[[1]])  # values
```

### Annotations

```{r annotations10}
a = annotations()
datatable(do.call(rbind, a))
```

## Advanced Usage




```{r mapping10}
a = mapping('projects')
a               # available information
a["defaults"]   # default fields
a[["defaults"]] # default fields as character()
a["fields"]     # all fields
```

# Advanced querying with filters

The GDC supports filtering based on fields and values. The GenomicDataCommons 
package allows writing queries using standard R syntax, simplifying the
creation of complicated filtering patterns. Examples follow:

```{r filtercases} 
youngcases = cases(filters=filters(diagnoses.age_at_diagnosis <= 10*365,endpoint='cases'))
```

# SessionInfo

```{r sessionInfo}
sessionInfo()
```
