---
title: "Introduction to Accessing the NCI Genomic Data Commons"
author: "Sean Davis & Martin Morgan"
date: "`r Sys.Date()`"
output:
  html_document:
    toc: true
    toc_float: true
    theme: united
    highlight: tango
    code_folding: hide
    df_print: paged
      
vignette: >
  %\VignetteIndexEntry{Introduction to Accessing the NCI Genomic Data Commons}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r init, results='hide', echo=FALSE, warning=FALSE, message=FALSE}
library(knitr)
```

```{r libraries, message=FALSE}
library(GenomicDataCommons)
library(DT)
```

# Background

From the [Genomic Data Commons (GDC) website](https://gdc.nci.nih.gov/about-gdc):

The National Cancer Institute's (NCI's) Genomic Data Commons (GDC) is
a data sharing platform that promotes precision medicine in
oncology. It is not just a database or a tool; it is an expandable
knowledge network supporting the import and standardization of genomic
and clinical data from cancer research programs.

The GDC contains NCI-generated data from some of the largest and most
comprehensive cancer genomic datasets, including The Cancer Genome
Atlas (TCGA) and Therapeutically Applicable Research to Generate
Effective Therapies (TARGET). For the first time, these datasets have
been harmonized using a common set of bioinformatics pipelines, so
that the data can be directly compared.

As a growing knowledge system for cancer, the GDC also enables
researchers to submit data, and harmonizes these data for import into
the GDC. As more researchers add clinical and genomic data to the GDC,
it will become an even more powerful tool for making discoveries about
the molecular basis of cancer that may lead to better care for
patients.

## Data Model

The
[data model for the GDC is complex](https://gdc.cancer.gov/developers/gdc-data-model/gdc-data-model-components),
but it worth a quick overview. The data model is encoded as a
so-called property graph. Nodes represent entities such as Projects,
Cases, Diagnoses, Files (various kinds), and Annotations. The
relationships between these entities are maintained as edges.  Both
nodes and edges may have Properties that supply instance details.  The
GDC API exposes these nodes and edges in a somewhat simplified set
of
[RESTful](https://en.wikipedia.org/wiki/Representational_state_transfer) endpoints.

## Full Feature Set of the GDC API

HTTP interface that uses JSON as the primary data exchange format

* Programmatic access to functionality provided by GDC Data and
  Submission portals, via projects, cases, files, annotations, data,
  slicing, status, and submission endpoints
* Programmatic access to GDC Legacy Archive via legacy endpoint
* Token-based authentication for secure access to controlled data and
  to submission functionality
* RESTful search that supports simple and complex queries via filters,
  fields, and facets parameters, and project, files, cases, and
  annotations endpoints.
* Search results can be sorted using sort parameter, paginated using
  size and from parameters, and output in JSON, TSV, and XML using
  format and pretty parameters.
* _mapping endpoint enables user discovery of fields available for
  data search and retrieval operations
* Support for downloading of individual files and of archives
  containing multiple files
* [NOT Implemented Here] Generation of download and upload manifests
  for use with the GDC Data Transfer Tool
* BAM slicing functionality for downloading part(s) of a BAM file
  specified using chromosomal coordinates or HGNC gene names
* [NOT Implemented Here] Transactional submission system that links
  individual data elements according to a graph-based GDC Data Model
* Two data entity identifiers: UUIDs, which are consistent across GDC,
  and Submitter IDs, for compatibility with submittersâ€™ tracking
  systems

Here, we will be describing how to access the main API endpoints:

- [Projects](https://docs.gdc.cancer.gov/Data_Portal/Users_Guide/Projects/)
- [Files](https://docs.gdc.cancer.gov/Data_Portal/Users_Guide/Cases_and_Files/)
- [Cases](https://docs.gdc.cancer.gov/Data_Portal/Users_Guide/Cases_and_Files/)
- [Annotations](https://docs.gdc.cancer.gov/Data_Portal/Users_Guide/Annotations/)
- [BAM Slicing](https://docs.gdc.cancer.gov/Data_Portal/Users_Guide/Cases_and_Files/#bam-slicing) - Specific to dealing with BAM files
- _mappings - For obtaining metadata about fields associated with the
  various endpoints

The following sections address each in turn, including examples of
usage.

# The API

## Basic API endpoints { .tabset .tabset-fade }

The Projects, Files, Cases, and Annotations endpoints are are all
accessed using functions of the same name (all lowercase). The default
behavior of each is to query the GDC and return the first 10 records
with a default set of fields. In this section, the results are simply
converted to a tabular form, but the data are not always tabular and
may have one-to-many relationships, so this represents a
simplification. Accessing the data more directly will be covered in
later sections.



### Projects

To get a sense of what is available in the default Project entities,
we simply call the `projects()` function without any arguments. By
default, the first 10 projects are returned. The returned value for
projects is a `data.frame` since the Project entity is tabular. 

```{r project10}
a = projects()
datatable(a)
```

### Files

```{r files10}
library(jsonlite)
fields <- c("data_type", "updated_datetime", "file_name", "data_format", "access",
            "data_category", "file_size", "type")
a = files(fields=fields)
a = files(fields=fields,filters=make_filter('data_format'=='VCF'))
# Convert to dataframe
files_df = do.call(rbind, a)
datatable(files_df[,c('file_name','data_format','access','type','data_category')],rownames=FALSE)
```

### Cases

```{r cases10}
a = cases()
a             # 10 cases by default
lengths(a)    # cases have different elements
a[1]          # summary
head(a[[1]])  # values
str(a[[1]])
```

### Annotations

Annotations are notes added to individual cases, samples or
files. They can be searched and filtered in the same way as other
entities. See
[here](https://docs.gdc.cancer.gov/Data_Portal/Users_Guide/Annotations/) for
details from Data Portal about annotations.

```{r annotations10}
a = annotations()
annots_df = do.call(rbind, a)
datatable(annots_df[,c('category','classification','notes','entity_type')],rownames=FALSE)
```

## Querying

Parameter | Default | Description
--------- | :-----: | ---------------------------------------------------
filters	| null | Specifies search parameters
format | JSON  | Specifies the API response format: JSON, XML, or TSV
pretty | false | Returns response with indentations and line breaks in a human-readable format
fields | null | Specifies which fields to include in the response
size | 10 | Specifies the number of results to return
from | 1 | Specifies the first record to return from a set of search results
sort | null | Specifies sorting for the search results
facets | null | Provides all existing values for a given field and the number of records having this value.

### Searching and filtering


Operator | Description | Number of Operands | Logic example
:---------:|----------------|:-------------:|-------------------------------
=  | equals (string or number) | one  | gender = "female"
!= | does not equal (string or number)  | one | project_id != "TARGET-AML"
<  | less than (number)  | one	| age at diagnosis < 90y
<=	| less than or equal (number)  | one | age at diagnosis <= 17
> | greater than (number)  | one | age at diagnosis > 50
>=	| greater than or equal (number) | one | age at diagnosis >= 18
is | is (missing) | one	| gender is missing
not	| not (missing)	| one | race not missing
in | matches a string or number in (a list)	| multiple | primary_site in [Brain, Lung]
exclude	| does not match any strings or values in (a list) | multiple | experimental_strategy exclude [WXS, WGS, "Genotyping array"]
and	| (operation1) and (operation2)	| multiple | {primary_site in [Brain, Lung]} and {gender = "female"}
or | (operation1) or (operation2) | multiple | {project_id != "TARGET-AML"} or {age at diagnosis < 90y}

### Faceting (counting)

```{r facetProjectByProgram}
res = cases(facets='diagnoses.tumor_stage')
```


### The _mapping endpoint

mapping key | Description of value
----------- | --------------------------------------------------
 \_mapping | All available fields and their descriptions. The endpoint-agnostic field names provided here are compatible with the filters parameter but are not always compatible with the fields parameter 
defaults | The default set of fields included in the API response when the fields parameter is not used in the request 
expand | Field group names for use with the expand parameter 
fields	| All available fields in an endpoint-specific format that is compatible with both the filters and fields parameters 
multi | GDC internal use 
nested  | Nested fields  




```{r mapping10}
a = mapping('projects')
colnames(a)
head(a)
# get the "default" fields
subset(a,defaults)
# get the "default" fields as character
subset(a,defaults)$field
```

# Advanced querying with filters

The GDC
supports
[advanced queries](https://docs.gdc.cancer.gov/Data_Portal/Users_Guide/Advanced_Search/) based
on filters and facets. The GenomicDataCommons package allows writing
queries using standard R syntax, simplifying the creation of
complicated filtering patterns. Examples follow:

```{r filtercases}
# youngcases = cases(filters=filters(diagnoses.age_at_diagnosis <= 10*365,endpoint='cases'))
```

# SessionInfo

```{r sessionInfo}
sessionInfo()
```
