---
title: "Working with somatic variants"
author:
- name: Sean Davis
  affiliation: National Cancer Institute, NIH
  email: seandavi@gmail.com
package: GenomicDataCommons
output: 
  BiocStyle::html_document2:
    toc_float: true
    theme: sandstone
vignette: >
  %\VignetteIndexEntry{Working with Somatic Variants}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}  
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, cache = TRUE)
```

# Preliminaries

```{r var_init}
library(knitr)
opts_chunk$set(cache=TRUE, message=FALSE, warning=TRUE)
```

```{r var_load_libraries}
library(GenomicDataCommons)
library(maftools)
library(SomaticSignatures)
```


# Variant resources at the GDC

Variant resources at the GDC come in two basic flavors, somatic and germline.
Germline variants are controlled-access data, requiring dbGaP approval for access.
Somatic variants are divided into two subclasses, masked and unmasked.
The somatic variants that are "masked" are available as open access data; 
"unmasked" variants are controlled access and require dbGaP access credentials.

```{r}
x = files() %>% 
    filter(~ data_category=='Simple Nucleotide Variation' & 
               data_format=='MAF' & 
               access=='open' & 
               analysis.workflow_type=='MuSE Variant Aggregation and Masking') %>% 
    results() %>% 
    as.data.frame()
```

