<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html>
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Find matching field names — grep_fields • GenomicDataCommons</title>

<!-- jquery -->
<script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script>
<!-- Bootstrap -->
<link href="https://maxcdn.bootstrapcdn.com/bootswatch/3.3.7/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">

<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/1.7.1/clipboard.min.js" integrity="sha384-cV+rhyOuRHc9Ub/91rihWcGmMmCXDeksTtCihMupQHSsi8GIIRDG0ThDc3HGQFJ3" crossorigin="anonymous"></script>

<!-- sticky kit -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>



<meta property="og:title" content="Find matching field names — grep_fields" />

<meta property="og:description" content="This utility function allows quick text-based search of available
fields for using grep" />
<meta name="twitter:card" content="summary" />



<!-- mathjax -->
<script src='https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML'></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


<!-- Global site tag (gtag.js) - Google Analytics -->
<script async src="https://www.googletagmanager.com/gtag/js?id=UA-93043521-1"></script>
<script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-93043521-1');
</script>

  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">GenomicDataCommons</a>
        <span class="label label-default" data-toggle="tooltip" data-placement="bottom" title="Released package">1.7.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/overview.html">The GenomicDataCommons Package</a>
    </li>
  </ul>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/Bioconductor/GenomicDataCommons">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Find matching field names</h1>
    <small class="dont-index">Source: <a href='https://github.com/Bioconductor/GenomicDataCommons/blob/master/R/fields.R'><code>R/fields.R</code></a></small>
    <div class="hidden name"><code>grep_fields.Rd</code></div>
    </div>

    <div class="ref-description">
    
    <p>This utility function allows quick text-based search of available
fields for using <code>grep</code></p>
    
    </div>

    <pre class="usage"><span class='fu'>grep_fields</span>(<span class='no'>entity</span>, <span class='no'>pattern</span>, <span class='no'>...</span>, <span class='kw'>value</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>)</pre>
    
    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>entity</th>
      <td><p>one of "files", "cases", "annotations", "projects"
against which to gather available fields for matching</p></td>
    </tr>
    <tr>
      <th>pattern</th>
      <td><p>A regular expression that will be used
in a call to <code>grep</code></p></td>
    </tr>
    <tr>
      <th>...</th>
      <td><p>passed on to grep</p></td>
    </tr>
    <tr>
      <th>value</th>
      <td><p>logical(1) whether to return values as opposed
to indices (passed along to grep)</p></td>
    </tr>
    </table>
    
    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>character() vector of field names matching
    <code>pattern</code></p>
    

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='fu'>grep_fields</span>(<span class='st'>'files'</span>,<span class='st'>'analysis'</span>)</div><div class='output co'>#&gt;   [1] "analysis.analysis_id"                                                     
#&gt;   [2] "analysis.analysis_type"                                                   
#&gt;   [3] "analysis.created_datetime"                                                
#&gt;   [4] "analysis.input_files.access"                                              
#&gt;   [5] "analysis.input_files.average_base_quality"                                
#&gt;   [6] "analysis.input_files.average_insert_size"                                 
#&gt;   [7] "analysis.input_files.average_read_length"                                 
#&gt;   [8] "analysis.input_files.created_datetime"                                    
#&gt;   [9] "analysis.input_files.data_category"                                       
#&gt;  [10] "analysis.input_files.data_format"                                         
#&gt;  [11] "analysis.input_files.data_type"                                           
#&gt;  [12] "analysis.input_files.error_type"                                          
#&gt;  [13] "analysis.input_files.experimental_strategy"                               
#&gt;  [14] "analysis.input_files.file_id"                                             
#&gt;  [15] "analysis.input_files.file_name"                                           
#&gt;  [16] "analysis.input_files.file_size"                                           
#&gt;  [17] "analysis.input_files.imaging_date"                                        
#&gt;  [18] "analysis.input_files.magnification"                                       
#&gt;  [19] "analysis.input_files.md5sum"                                              
#&gt;  [20] "analysis.input_files.mean_coverage"                                       
#&gt;  [21] "analysis.input_files.pairs_on_diff_chr"                                   
#&gt;  [22] "analysis.input_files.platform"                                            
#&gt;  [23] "analysis.input_files.proportion_base_mismatch"                            
#&gt;  [24] "analysis.input_files.proportion_coverage_10X"                             
#&gt;  [25] "analysis.input_files.proportion_coverage_30X"                             
#&gt;  [26] "analysis.input_files.proportion_reads_duplicated"                         
#&gt;  [27] "analysis.input_files.proportion_reads_mapped"                             
#&gt;  [28] "analysis.input_files.proportion_targets_no_coverage"                      
#&gt;  [29] "analysis.input_files.read_pair_number"                                    
#&gt;  [30] "analysis.input_files.revision"                                            
#&gt;  [31] "analysis.input_files.state"                                               
#&gt;  [32] "analysis.input_files.state_comment"                                       
#&gt;  [33] "analysis.input_files.submitter_id"                                        
#&gt;  [34] "analysis.input_files.total_reads"                                         
#&gt;  [35] "analysis.input_files.updated_datetime"                                    
#&gt;  [36] "analysis.metadata.read_groups.adapter_name"                               
#&gt;  [37] "analysis.metadata.read_groups.adapter_sequence"                           
#&gt;  [38] "analysis.metadata.read_groups.base_caller_name"                           
#&gt;  [39] "analysis.metadata.read_groups.base_caller_version"                        
#&gt;  [40] "analysis.metadata.read_groups.created_datetime"                           
#&gt;  [41] "analysis.metadata.read_groups.days_to_sequencing"                         
#&gt;  [42] "analysis.metadata.read_groups.experiment_name"                            
#&gt;  [43] "analysis.metadata.read_groups.flow_cell_barcode"                          
#&gt;  [44] "analysis.metadata.read_groups.fragment_maximum_length"                    
#&gt;  [45] "analysis.metadata.read_groups.fragment_mean_length"                       
#&gt;  [46] "analysis.metadata.read_groups.fragment_minimum_length"                    
#&gt;  [47] "analysis.metadata.read_groups.fragment_standard_deviation_length"         
#&gt;  [48] "analysis.metadata.read_groups.includes_spike_ins"                         
#&gt;  [49] "analysis.metadata.read_groups.instrument_model"                           
#&gt;  [50] "analysis.metadata.read_groups.is_paired_end"                              
#&gt;  [51] "analysis.metadata.read_groups.lane_number"                                
#&gt;  [52] "analysis.metadata.read_groups.library_name"                               
#&gt;  [53] "analysis.metadata.read_groups.library_preparation_kit_catalog_number"     
#&gt;  [54] "analysis.metadata.read_groups.library_preparation_kit_name"               
#&gt;  [55] "analysis.metadata.read_groups.library_preparation_kit_vendor"             
#&gt;  [56] "analysis.metadata.read_groups.library_preparation_kit_version"            
#&gt;  [57] "analysis.metadata.read_groups.library_selection"                          
#&gt;  [58] "analysis.metadata.read_groups.library_strand"                             
#&gt;  [59] "analysis.metadata.read_groups.library_strategy"                           
#&gt;  [60] "analysis.metadata.read_groups.multiplex_barcode"                          
#&gt;  [61] "analysis.metadata.read_groups.platform"                                   
#&gt;  [62] "analysis.metadata.read_groups.read_group_id"                              
#&gt;  [63] "analysis.metadata.read_groups.read_group_name"                            
#&gt;  [64] "analysis.metadata.read_groups.read_group_qcs.adapter_content"             
#&gt;  [65] "analysis.metadata.read_groups.read_group_qcs.basic_statistics"            
#&gt;  [66] "analysis.metadata.read_groups.read_group_qcs.created_datetime"            
#&gt;  [67] "analysis.metadata.read_groups.read_group_qcs.encoding"                    
#&gt;  [68] "analysis.metadata.read_groups.read_group_qcs.fastq_name"                  
#&gt;  [69] "analysis.metadata.read_groups.read_group_qcs.kmer_content"                
#&gt;  [70] "analysis.metadata.read_groups.read_group_qcs.overrepresented_sequences"   
#&gt;  [71] "analysis.metadata.read_groups.read_group_qcs.per_base_n_content"          
#&gt;  [72] "analysis.metadata.read_groups.read_group_qcs.per_base_sequence_content"   
#&gt;  [73] "analysis.metadata.read_groups.read_group_qcs.per_base_sequence_quality"   
#&gt;  [74] "analysis.metadata.read_groups.read_group_qcs.per_sequence_gc_content"     
#&gt;  [75] "analysis.metadata.read_groups.read_group_qcs.per_sequence_quality_score"  
#&gt;  [76] "analysis.metadata.read_groups.read_group_qcs.per_tile_sequence_quality"   
#&gt;  [77] "analysis.metadata.read_groups.read_group_qcs.percent_gc_content"          
#&gt;  [78] "analysis.metadata.read_groups.read_group_qcs.read_group_qc_id"            
#&gt;  [79] "analysis.metadata.read_groups.read_group_qcs.sequence_duplication_levels" 
#&gt;  [80] "analysis.metadata.read_groups.read_group_qcs.sequence_length_distribution"
#&gt;  [81] "analysis.metadata.read_groups.read_group_qcs.state"                       
#&gt;  [82] "analysis.metadata.read_groups.read_group_qcs.submitter_id"                
#&gt;  [83] "analysis.metadata.read_groups.read_group_qcs.total_sequences"             
#&gt;  [84] "analysis.metadata.read_groups.read_group_qcs.updated_datetime"            
#&gt;  [85] "analysis.metadata.read_groups.read_group_qcs.workflow_end_datetime"       
#&gt;  [86] "analysis.metadata.read_groups.read_group_qcs.workflow_link"               
#&gt;  [87] "analysis.metadata.read_groups.read_group_qcs.workflow_start_datetime"     
#&gt;  [88] "analysis.metadata.read_groups.read_group_qcs.workflow_type"               
#&gt;  [89] "analysis.metadata.read_groups.read_group_qcs.workflow_version"            
#&gt;  [90] "analysis.metadata.read_groups.read_length"                                
#&gt;  [91] "analysis.metadata.read_groups.RIN"                                        
#&gt;  [92] "analysis.metadata.read_groups.sequencing_center"                          
#&gt;  [93] "analysis.metadata.read_groups.sequencing_date"                            
#&gt;  [94] "analysis.metadata.read_groups.size_selection_range"                       
#&gt;  [95] "analysis.metadata.read_groups.spike_ins_concentration"                    
#&gt;  [96] "analysis.metadata.read_groups.spike_ins_fasta"                            
#&gt;  [97] "analysis.metadata.read_groups.state"                                      
#&gt;  [98] "analysis.metadata.read_groups.submitter_id"                               
#&gt;  [99] "analysis.metadata.read_groups.target_capture_kit"                         
#&gt; [100] "analysis.metadata.read_groups.target_capture_kit_catalog_number"          
#&gt; [101] "analysis.metadata.read_groups.target_capture_kit_name"                    
#&gt; [102] "analysis.metadata.read_groups.target_capture_kit_target_region"           
#&gt; [103] "analysis.metadata.read_groups.target_capture_kit_vendor"                  
#&gt; [104] "analysis.metadata.read_groups.target_capture_kit_version"                 
#&gt; [105] "analysis.metadata.read_groups.to_trim_adapter_sequence"                   
#&gt; [106] "analysis.metadata.read_groups.updated_datetime"                           
#&gt; [107] "analysis.state"                                                           
#&gt; [108] "analysis.submitter_id"                                                    
#&gt; [109] "analysis.updated_datetime"                                                
#&gt; [110] "analysis.workflow_end_datetime"                                           
#&gt; [111] "analysis.workflow_link"                                                   
#&gt; [112] "analysis.workflow_start_datetime"                                         
#&gt; [113] "analysis.workflow_type"                                                   
#&gt; [114] "analysis.workflow_version"                                                
#&gt; [115] "downstream_analyses.analysis_id"                                          
#&gt; [116] "downstream_analyses.analysis_type"                                        </div><div class='input'>
</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      
      <li><a href="#value">Value</a></li>
      
      <li><a href="#examples">Examples</a></li>
    </ul>

  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by Martin Morgan, Sean Davis.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://pkgdown.r-lib.org/">pkgdown</a>.</p>
</div>

      </footer>
   </div>

  

  </body>
</html>

