<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Using the GDC API • GenomicDataCommons</title>
<!-- jquery --><script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script><!-- Font Awesome icons --><link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">
<!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../jquery.sticky-kit.min.js"></script><script src="../pkgdown.js"></script><!-- mathjax --><script src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-vignette">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="../index.html">GenomicDataCommons</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../articles/index.html">Articles</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><script src="api_files/htmlwidgets-0.8/htmlwidgets.js"></script><script src="api_files/jquery-1.12.4/jquery.min.js"></script><script src="api_files/datatables-binding-0.2/datatables.js"></script><link href="api_files/dt-core-1.10.12/css/jquery.dataTables.min.css" rel="stylesheet">
<link href="api_files/dt-core-1.10.12/css/jquery.dataTables.extra.css" rel="stylesheet">
<script src="api_files/dt-core-1.10.12/js/jquery.dataTables.min.js"></script><div class="row">
  <div class="col-md-9">
    <div class="page-header toc-ignore">
      <h1>Using the GDC API</h1>
                        <h4 class="author">Sean Davis &amp; Martin Morgan</h4>
            
            <h4 class="date">2017-01-25</h4>
          </div>

    
    
<div class="contents">
<div id="the-api" class="section level1 tabset tabset-fade">
<h1 class="hasAnchor">
<html><body><a href="#the-api" class="anchor"> </a></body></html>The API</h1>
<div id="querying-metadata" class="section level2 tabset tabset-fade">
<h2 class="hasAnchor">
<html><body><a href="#querying-metadata" class="anchor"> </a></body></html>Querying Metadata</h2>
<div id="general-concepts" class="section level3">
<h3 class="hasAnchor">
<html><body><a href="#general-concepts" class="anchor"> </a></body></html>General Concepts</h3>
<p>The Projects, Files, Cases, and Annotations endpoints are are all accessed using functions of the same name (all lowercase). The default behavior of each is to query the GDC and return the first 10 records with a default set of fields. In this section, the results are simply converted to a tabular form, but the data are not always tabular and may have one-to-many relationships, so this represents a simplification. Accessing the data more directly will be covered in later sections.</p>
<div id="querying" class="section level4">
<h4 class="hasAnchor">
<html><body><a href="#querying" class="anchor"> </a></body></html>Querying</h4>
<table style="width:97%;" class="table">
<colgroup>
<col width="13%">
<col width="11%">
<col width="72%">
</colgroup>
<thead><tr class="header">
<th align="left">Parameter</th>
<th align="center">Default</th>
<th align="left">Description</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">filters</td>
<td align="center">null</td>
<td align="left">Specifies search parameters</td>
</tr>
<tr class="even">
<td align="left">format</td>
<td align="center">JSON</td>
<td align="left">Specifies the API response format: JSON, XML, or TSV</td>
</tr>
<tr class="odd">
<td align="left">pretty</td>
<td align="center">false</td>
<td align="left">Returns response with indentations and line breaks in a human-readable format</td>
</tr>
<tr class="even">
<td align="left">fields</td>
<td align="center">null</td>
<td align="left">Specifies which fields to include in the response</td>
</tr>
<tr class="odd">
<td align="left">size</td>
<td align="center">10</td>
<td align="left">Specifies the number of results to return</td>
</tr>
<tr class="even">
<td align="left">from</td>
<td align="center">1</td>
<td align="left">Specifies the first record to return from a set of search results</td>
</tr>
<tr class="odd">
<td align="left">sort</td>
<td align="center">null</td>
<td align="left">Specifies sorting for the search results</td>
</tr>
<tr class="even">
<td align="left">facets</td>
<td align="center">null</td>
<td align="left">Provides all existing values for a given field and the number of records having this value.</td>
</tr>
</tbody>
</table>
</div>
<div id="searching-and-filtering" class="section level4">
<h4 class="hasAnchor">
<html><body><a href="#searching-and-filtering" class="anchor"> </a></body></html>Searching and filtering</h4>
<table style="width:107%;" class="table">
<colgroup>
<col width="16%">
<col width="23%">
<col width="22%">
<col width="44%">
</colgroup>
<thead><tr class="header">
<th align="center">Operator</th>
<th align="left">Description</th>
<th align="center">Number of Operands</th>
<th align="left">Logic example</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="center">=</td>
<td align="left">equals (string or number)</td>
<td align="center">one</td>
<td align="left">gender = “female”</td>
</tr>
<tr class="even">
<td align="center">!=</td>
<td align="left">does not equal (string or number)</td>
<td align="center">one</td>
<td align="left">project_id != “TARGET-AML”</td>
</tr>
<tr class="odd">
<td align="center">&lt;</td>
<td align="left">less than (number)</td>
<td align="center">one</td>
<td align="left">age at diagnosis &lt; 90y</td>
</tr>
<tr class="even">
<td align="center">&lt;=</td>
<td align="left">less than or equal (number)</td>
<td align="center">one</td>
<td align="left">age at diagnosis &lt;= 17</td>
</tr>
<tr class="odd">
<td align="center">&gt;</td>
<td align="left">greater than (number)</td>
<td align="center">one</td>
<td align="left">age at diagnosis &gt; 50</td>
</tr>
<tr class="even">
<td align="center">&gt;=</td>
<td align="left">greater than or equal (number)</td>
<td align="center">one</td>
<td align="left">age at diagnosis &gt;= 18</td>
</tr>
<tr class="odd">
<td align="center">is</td>
<td align="left">is (missing)</td>
<td align="center">one</td>
<td align="left">gender is missing</td>
</tr>
<tr class="even">
<td align="center">not</td>
<td align="left">not (missing)</td>
<td align="center">one</td>
<td align="left">race not missing</td>
</tr>
<tr class="odd">
<td align="center">in</td>
<td align="left">matches a string or number in (a list)</td>
<td align="center">multiple</td>
<td align="left">primary_site in [Brain, Lung]</td>
</tr>
<tr class="even">
<td align="center">exclude</td>
<td align="left">does not match any strings or values in (a list)</td>
<td align="center">multiple</td>
<td align="left">experimental_strategy exclude [WXS, WGS, “Genotyping array”]</td>
</tr>
<tr class="odd">
<td align="center">and</td>
<td align="left">(operation1) and (operation2)</td>
<td align="center">multiple</td>
<td align="left">{primary_site in [Brain, Lung]} and {gender = “female”}</td>
</tr>
<tr class="even">
<td align="center">or</td>
<td align="left">(operation1) or (operation2)</td>
<td align="center">multiple</td>
<td align="left">{project_id != “TARGET-AML”} or {age at diagnosis &lt; 90y}</td>
</tr>
</tbody>
</table>
</div>
<div id="the-mapping-endpoint" class="section level4">
<h4 class="hasAnchor">
<html><body><a href="#the-mapping-endpoint" class="anchor"> </a></body></html>The mapping endpoint</h4>
<p>The mapping endpoint returns the available fields and the types of data contained in them for each of the four primary metadata endpoints. A data frame with is returned when the <code><a href="../reference/mapping.html">mapping()</a></code> function is called with one of “cases”, “files”, “projects”, or “annotations”. In addition to the column definitions (first 5 columns), four additional columns are returned that allow subsetting fields to those of interest. The endpoint-agnostic field names provided by the mapping endpoint are compatible with the filters parameter but are not always compatible with the fields parameter.</p>
<table style="width:88%;" class="table">
<colgroup>
<col width="16%">
<col width="70%">
</colgroup>
<thead><tr class="header">
<th align="left">mapping key</th>
<th align="left">Description of value</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">defaults</td>
<td align="left">The default set of fields included in the API response when the fields parameter is not used in the request</td>
</tr>
<tr class="even">
<td align="left">expand</td>
<td align="left">Field group names for use with the expand parameter</td>
</tr>
<tr class="odd">
<td align="left">fields</td>
<td align="left">All available fields in an endpoint-specific format that is compatible with both the filters and fields parameters</td>
</tr>
<tr class="even">
<td align="left">multi</td>
<td align="left">GDC internal use</td>
</tr>
<tr class="odd">
<td align="left">nested</td>
<td align="left">Nested fields</td>
</tr>
</tbody>
</table>
<p>As an example, we can look at the fields associated with the <a href="#files">files</a> endpoint.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">a =<span class="st"> </span><span class="kw"><a href="../reference/mapping.html">mapping</a></span>(<span class="st">'files'</span>)
<span class="kw">colnames</span>(a)</code></pre></div>
<pre><code>## [1] "description" "doc_type"    "field"       "full"        "type"       
## [6] "defaults"    "expand"      "multi"       "nested"</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">head</span>(a)</code></pre></div>
<div data-pagedtable="false">
<script data-pagedtable-source type="application/json">
{"columns":[{"label":["description"],"name":[1],"type":["fctr"],"align":["left"]},{"label":["doc_type"],"name":[2],"type":["fctr"],"align":["left"]},{"label":["field"],"name":[3],"type":["fctr"],"align":["left"]},{"label":["full"],"name":[4],"type":["fctr"],"align":["left"]},{"label":["type"],"name":[5],"type":["fctr"],"align":["left"]},{"label":["defaults"],"name":[6],"type":["lgl"],"align":["right"]},{"label":["expand"],"name":[7],"type":["lgl"],"align":["right"]},{"label":["multi"],"name":[8],"type":["lgl"],"align":["right"]},{"label":["nested"],"name":[9],"type":["lgl"],"align":["right"]}],"data":[{"1":"","2":"files","3":"access","4":"files.access","5":"string","6":"TRUE","7":"FALSE","8":"FALSE","9":"FALSE"},{"1":"","2":"files","3":"acl","4":"files.acl","5":"string","6":"TRUE","7":"FALSE","8":"FALSE","9":"FALSE"},{"1":"","2":"files","3":"analysis.analysis_id","4":"files.analysis.analysis_id","5":"string","6":"FALSE","7":"FALSE","8":"FALSE","9":"FALSE"},{"1":"","2":"files","3":"analysis.analysis_type","4":"files.analysis.analysis_type","5":"string","6":"FALSE","7":"FALSE","8":"FALSE","9":"FALSE"},{"1":"","2":"files","3":"analysis.created_datetime","4":"files.analysis.created_datetime","5":"string","6":"FALSE","7":"FALSE","8":"FALSE","9":"FALSE"},{"1":"","2":"files","3":"analysis.input_files.access","4":"files.analysis.input_files.access","5":"string","6":"FALSE","7":"FALSE","8":"FALSE","9":"FALSE"}],"options":{"columns":{"min":{},"max":[10]},"rows":{"min":[10],"max":[10]},"pages":{}}}
  </script>
</div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># get the "default" fields</span>
<span class="kw">subset</span>(a,defaults)</code></pre></div>
<div data-pagedtable="false">
<script data-pagedtable-source type="application/json">
{"columns":[{"label":["description"],"name":[1],"type":["fctr"],"align":["left"]},{"label":["doc_type"],"name":[2],"type":["fctr"],"align":["left"]},{"label":["field"],"name":[3],"type":["fctr"],"align":["left"]},{"label":["full"],"name":[4],"type":["fctr"],"align":["left"]},{"label":["type"],"name":[5],"type":["fctr"],"align":["left"]},{"label":["defaults"],"name":[6],"type":["lgl"],"align":["right"]},{"label":["expand"],"name":[7],"type":["lgl"],"align":["right"]},{"label":["multi"],"name":[8],"type":["lgl"],"align":["right"]},{"label":["nested"],"name":[9],"type":["lgl"],"align":["right"]}],"data":[{"1":"","2":"files","3":"access","4":"files.access","5":"string","6":"TRUE","7":"FALSE","8":"FALSE","9":"FALSE"},{"1":"","2":"files","3":"acl","4":"files.acl","5":"string","6":"TRUE","7":"FALSE","8":"FALSE","9":"FALSE"},{"1":"","2":"files","3":"created_datetime","4":"files.created_datetime","5":"string","6":"TRUE","7":"FALSE","8":"FALSE","9":"FALSE"},{"1":"","2":"files","3":"data_category","4":"files.data_category","5":"string","6":"TRUE","7":"FALSE","8":"FALSE","9":"FALSE"},{"1":"","2":"files","3":"data_format","4":"files.data_format","5":"string","6":"TRUE","7":"FALSE","8":"FALSE","9":"FALSE"},{"1":"","2":"files","3":"data_type","4":"files.data_type","5":"string","6":"TRUE","7":"FALSE","8":"FALSE","9":"FALSE"},{"1":"","2":"files","3":"error_type","4":"files.error_type","5":"string","6":"TRUE","7":"FALSE","8":"FALSE","9":"FALSE"},{"1":"","2":"files","3":"experimental_strategy","4":"files.experimental_strategy","5":"string","6":"TRUE","7":"FALSE","8":"FALSE","9":"FALSE"},{"1":"","2":"files","3":"file_id","4":"files.file_id","5":"id","6":"TRUE","7":"FALSE","8":"TRUE","9":"FALSE"},{"1":"","2":"files","3":"file_name","4":"files.file_name","5":"id","6":"TRUE","7":"FALSE","8":"TRUE","9":"FALSE"},{"1":"","2":"files","3":"file_size","4":"files.file_size","5":"long","6":"TRUE","7":"FALSE","8":"FALSE","9":"FALSE"},{"1":"","2":"files","3":"file_state","4":"files.file_state","5":"string","6":"TRUE","7":"FALSE","8":"FALSE","9":"FALSE"},{"1":"","2":"files","3":"md5sum","4":"files.md5sum","5":"string","6":"TRUE","7":"FALSE","8":"FALSE","9":"FALSE"},{"1":"","2":"files","3":"origin","4":"files.origin","5":"string","6":"TRUE","7":"FALSE","8":"FALSE","9":"FALSE"},{"1":"","2":"files","3":"platform","4":"files.platform","5":"string","6":"TRUE","7":"FALSE","8":"FALSE","9":"FALSE"},{"1":"","2":"files","3":"revision","4":"files.revision","5":"long","6":"TRUE","7":"FALSE","8":"FALSE","9":"FALSE"},{"1":"","2":"files","3":"state","4":"files.state","5":"string","6":"TRUE","7":"FALSE","8":"FALSE","9":"FALSE"},{"1":"","2":"files","3":"state_comment","4":"files.state_comment","5":"string","6":"TRUE","7":"FALSE","8":"FALSE","9":"FALSE"},{"1":"","2":"files","3":"submitter_id","4":"files.submitter_id","5":"string","6":"TRUE","7":"FALSE","8":"FALSE","9":"FALSE"},{"1":"","2":"files","3":"tags","4":"files.tags","5":"string","6":"TRUE","7":"FALSE","8":"FALSE","9":"FALSE"},{"1":"","2":"files","3":"type","4":"files.type","5":"string","6":"TRUE","7":"FALSE","8":"FALSE","9":"FALSE"},{"1":"","2":"files","3":"updated_datetime","4":"files.updated_datetime","5":"string","6":"TRUE","7":"FALSE","8":"FALSE","9":"FALSE"}],"options":{"columns":{"min":{},"max":[10]},"rows":{"min":[10],"max":[10]},"pages":{}}}
  </script>
</div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># get the "default" fields as character</span>
<span class="kw">subset</span>(a,defaults)$field</code></pre></div>
<pre><code>##  [1] access                acl                   created_datetime     
##  [4] data_category         data_format           data_type            
##  [7] error_type            experimental_strategy file_id              
## [10] file_name             file_size             file_state           
## [13] md5sum                origin                platform             
## [16] revision              state                 state_comment        
## [19] submitter_id          tags                  type                 
## [22] updated_datetime     
## 221 Levels: access acl analysis.analysis_id ... updated_datetime</code></pre>
</div>
<div id="faceting-counting" class="section level4">
<h4 class="hasAnchor">
<html><body><a href="#faceting-counting" class="anchor"> </a></body></html>Faceting (counting)</h4>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">res =<span class="st"> </span><span class="kw"><a href="../reference/query.html">cases</a></span>(<span class="dt">facets=</span><span class="st">'diagnoses.tumor_stage'</span>)</code></pre></div>
</div>
</div>
<div id="projects" class="section level3">
<h3 class="hasAnchor">
<html><body><a href="#projects" class="anchor"> </a></body></html>Projects</h3>
<p>To get a sense of what is available in the default Project entities, we simply call the <code><a href="../reference/query.html">projects()</a></code> function without any arguments. By default, the first 10 projects are returned. The returned value for projects is a <code>data.frame</code> since the Project entity is tabular.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">a =<span class="st"> </span><span class="kw"><a href="../reference/query.html">projects</a></span>()</code></pre></div>
<pre><code>## No encoding supplied: defaulting to UTF-8.</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">datatable</span>(a)</code></pre></div>
<div id="htmlwidget-6d217269fc06ad440264" style="width:100%;height:auto;" class="datatables html-widget"></div>
<script type="application/json" data-for="htmlwidget-6d217269fc06ad440264">{"x":{"filter":"none","data":[["1","2","3","4","5","6","7","8","9","10"],["","","phs000467","","","","","","","phs000468"],["Sarcoma","Pancreatic Adenocarcinoma","Neuroblastoma","Rectum Adenocarcinoma","Mesothelioma","Glioblastoma Multiforme","Kidney Renal Papillary Cell Carcinoma","Adrenocortical Carcinoma","Cervical Squamous Cell Carcinoma and Endocervical Adenocarcinoma","Osteosarcoma"],["true","true","true","true","true","true","true","true","true","true"],["legacy","legacy","legacy","legacy","legacy","legacy","legacy","legacy","legacy","legacy"],["Soft Tissue","Pancreas","Nervous System","Colorectal","Pleura","Brain","Kidney","Adrenal Gland","Cervix","Bone"],["TCGA-SARC","TCGA-PAAD","TARGET-NBL","TCGA-READ","TCGA-MESO","TCGA-GBM","TCGA-KIRP","TCGA-ACC","TCGA-CESC","TARGET-OS"],["Sarcoma","Pancreatic Adenocarcinoma","Neuroblastoma","Rectum Adenocarcinoma","Mesothelioma","Glioblastoma Multiforme","Kidney Renal Papillary Cell Carcinoma","Adrenocortical Carcinoma","Cervical Squamous Cell Carcinoma and Endocervical Adenocarcinoma","Osteosarcoma"]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th> \u003c/th>\n      <th>dbgap_accession_number\u003c/th>\n      <th>disease_type\u003c/th>\n      <th>released\u003c/th>\n      <th>state\u003c/th>\n      <th>primary_site\u003c/th>\n      <th>project_id\u003c/th>\n      <th>name\u003c/th>\n    \u003c/tr>\n  \u003c/thead>\n\u003c/table>","options":{"order":[],"autoWidth":false,"orderClasses":false,"columnDefs":[{"orderable":false,"targets":0}]}},"evals":[],"jsHooks":[]}</script>
</div>
<div id="files" class="section level3">
<h3 class="hasAnchor">
<html><body><a href="#files" class="anchor"> </a></body></html>Files</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(jsonlite)
fields &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">"data_type"</span>, <span class="st">"updated_datetime"</span>, <span class="st">"file_name"</span>, <span class="st">"data_format"</span>, <span class="st">"access"</span>,
            <span class="st">"data_category"</span>, <span class="st">"file_size"</span>, <span class="st">"type"</span>)
a =<span class="st"> </span><span class="kw"><a href="../reference/query.html">files</a></span>(<span class="dt">fields=</span>fields)
a =<span class="st"> </span><span class="kw"><a href="../reference/query.html">files</a></span>(<span class="dt">fields=</span>fields,<span class="dt">filters=</span><span class="kw"><a href="../reference/make_filter.html">make_filter</a></span>(<span class="st">'data_format'</span>==<span class="st">'VCF'</span>))
<span class="co"># Convert to dataframe</span>
files_df =<span class="st"> </span><span class="kw">do.call</span>(rbind, a)
<span class="kw">datatable</span>(files_df[,<span class="kw">c</span>(<span class="st">'file_name'</span>,<span class="st">'data_format'</span>,<span class="st">'access'</span>,<span class="st">'type'</span>,<span class="st">'data_category'</span>)],<span class="dt">rownames=</span><span class="ot">FALSE</span>)</code></pre></div>
<div id="htmlwidget-8354599c29dad4a8ce77" style="width:100%;height:auto;" class="datatables html-widget"></div>
<script type="application/json" data-for="htmlwidget-8354599c29dad4a8ce77">{"x":{"filter":"none","data":[["8bfeb962-e2b3-463a-9fbb-f5c6296f3ec2.vep.reheader.vcf.gz","796ea613-f0da-4592-a29e-74edf27a442a.vcf.gz","4dd5ba61-80ac-440f-aced-ce0d3d468f6b.vep.reheader.vcf.gz","2a84193a-3d0a-4b2d-a53c-a2a98ba09ab6.vep.reheader.vcf.gz","324b7742-65bf-414a-b0c1-8743c6e0973b.vcf.gz","3da4e2e4-bc5f-4a77-acc4-acd81d1dc85a.vep.reheader.vcf.gz","8c56919e-537f-40e8-adf7-392f395547ce.vep.reheader.vcf.gz","ed457ca4-4475-4cbd-98ce-07de9c887ed8.vep.reheader.vcf.gz","9cb1ccd9-76df-4014-af33-60639229b63b.vcf.gz","5f6a3b19-750d-4868-9e97-e057a1c5dd0a.vcf.gz"],["VCF","VCF","VCF","VCF","VCF","VCF","VCF","VCF","VCF","VCF"],["controlled","controlled","controlled","controlled","controlled","controlled","controlled","controlled","controlled","controlled"],["annotated_somatic_mutation","simple_somatic_mutation","annotated_somatic_mutation","annotated_somatic_mutation","simple_somatic_mutation","annotated_somatic_mutation","annotated_somatic_mutation","annotated_somatic_mutation","simple_somatic_mutation","simple_somatic_mutation"],["Simple Nucleotide Variation","Simple Nucleotide Variation","Simple Nucleotide Variation","Simple Nucleotide Variation","Simple Nucleotide Variation","Simple Nucleotide Variation","Simple Nucleotide Variation","Simple Nucleotide Variation","Simple Nucleotide Variation","Simple Nucleotide Variation"]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th>file_name\u003c/th>\n      <th>data_format\u003c/th>\n      <th>access\u003c/th>\n      <th>type\u003c/th>\n      <th>data_category\u003c/th>\n    \u003c/tr>\n  \u003c/thead>\n\u003c/table>","options":{"order":[],"autoWidth":false,"orderClasses":false}},"evals":[],"jsHooks":[]}</script>
</div>
<div id="cases" class="section level3">
<h3 class="hasAnchor">
<html><body><a href="#cases" class="anchor"> </a></body></html>Cases</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">a =<span class="st"> </span><span class="kw"><a href="../reference/query.html">cases</a></span>()
a             <span class="co"># 10 cases by default</span></code></pre></div>
<pre><code>## class: cases_list
## cases: 10
## names:
##     5f468c25-73c3-49d5-ba12-50653dfe2c41,
##     52c35691-efca-42f5-a03e-02c9f872275c,
##     5f99343b-e8ea-4b41-b36d-cd5c8fd918d0, ...,
##     643ac1e9-3b8d-5cfa-ab52-18de4bc8d3d3,
##     fcf64b55-5c4f-4d82-ac6a-4713d01143cb</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">lengths</span>(a)    <span class="co"># cases have different elements</span></code></pre></div>
<pre><code>## 5f468c25-73c3-49d5-ba12-50653dfe2c41 52c35691-efca-42f5-a03e-02c9f872275c 
##                                   13                                   13 
## 5f99343b-e8ea-4b41-b36d-cd5c8fd918d0 fafd6f5b-1d76-4537-bd1c-e0bd7b4e2166 
##                                   13                                   13 
## 183dd089-e932-4be2-b252-0e8572e7da4e 9283709d-d019-4291-a097-611a5c05244b 
##                                   13                                   13 
## 89c72f0e-e724-4826-bb9c-d7524d1b1258 929837f9-297a-47c0-ac9a-a142eef620d3 
##                                   13                                   13 
## 643ac1e9-3b8d-5cfa-ab52-18de4bc8d3d3 fcf64b55-5c4f-4d82-ac6a-4713d01143cb 
##                                    6                                   13</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">a[<span class="dv">1</span>]          <span class="co"># summary</span></code></pre></div>
<pre><code>## class: cases_list
## cases: 1
## names:
##     5f468c25-73c3-49d5-ba12-50653dfe2c41
## fields:
##     sample_ids(2)
##     portion_ids(2)
##     submitter_portion_ids(2)
##     created_datetime(0)
##     submitter_aliquot_ids(12)
##     updated_datetime(1)
##     submitter_analyte_ids(5)
##     analyte_ids(5)
##     submitter_id(1)
##     state(0)
##     aliquot_ids(12)
##     slide_ids(1)
##     submitter_sample_ids(2)</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">head</span>(a[[<span class="dv">1</span>]])  <span class="co"># values</span></code></pre></div>
<pre><code>## $sample_ids
## [1] "5d809b44-5c5e-4942-ac8e-a5650e1e1ccc"
## [2] "f80e9141-3634-49cd-8c85-610c7f644d83"
## 
## $portion_ids
## [1] "08d8cb7b-1725-4f3e-8e16-989a232de883"
## [2] "35e98f3f-fa01-4b37-9892-7292aa908b6f"
## 
## $submitter_portion_ids
## [1] "TCGA-3A-A9IJ-01A-11" "TCGA-3A-A9IJ-10A-01"
## 
## $created_datetime
## NULL
## 
## $submitter_aliquot_ids
##  [1] "TCGA-3A-A9IJ-01A-11D-A398-05" "TCGA-3A-A9IJ-01A-11R-A39J-13"
##  [3] "TCGA-3A-A9IJ-01A-11D-A395-26" "TCGA-3A-A9IJ-01A-11R-A39D-07"
##  [5] "TCGA-3A-A9IJ-10A-01D-A395-26" "TCGA-3A-A9IJ-10A-01D-A399-01"
##  [7] "TCGA-3A-A9IJ-10A-01D-A39A-08" "TCGA-3A-A9IJ-01A-01D-YYYY-23"
##  [9] "TCGA-3A-A9IJ-01A-11D-A396-01" "TCGA-3A-A9IJ-01A-11W-A450-08"
## [11] "TCGA-3A-A9IJ-01A-11D-A397-08" "TCGA-3A-A9IJ-10A-01W-A451-08"
## 
## $updated_datetime
## [1] "2016-09-08T10:25:08.392801-05:00"</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">str</span>(a[[<span class="dv">1</span>]])</code></pre></div>
<pre><code>## List of 13
##  $ sample_ids           : chr [1:2] "5d809b44-5c5e-4942-ac8e-a5650e1e1ccc" "f80e9141-3634-49cd-8c85-610c7f644d83"
##  $ portion_ids          : chr [1:2] "08d8cb7b-1725-4f3e-8e16-989a232de883" "35e98f3f-fa01-4b37-9892-7292aa908b6f"
##  $ submitter_portion_ids: chr [1:2] "TCGA-3A-A9IJ-01A-11" "TCGA-3A-A9IJ-10A-01"
##  $ created_datetime     : NULL
##  $ submitter_aliquot_ids: chr [1:12] "TCGA-3A-A9IJ-01A-11D-A398-05" "TCGA-3A-A9IJ-01A-11R-A39J-13" "TCGA-3A-A9IJ-01A-11D-A395-26" "TCGA-3A-A9IJ-01A-11R-A39D-07" ...
##  $ updated_datetime     : chr "2016-09-08T10:25:08.392801-05:00"
##  $ submitter_analyte_ids: chr [1:5] "TCGA-3A-A9IJ-01A-11R" "TCGA-3A-A9IJ-01A-11D" "TCGA-3A-A9IJ-10A-01D" "TCGA-3A-A9IJ-10A-01W" ...
##  $ analyte_ids          : chr [1:5] "57518237-1d1e-4c17-967f-3de1ddff7e50" "e995c697-d0de-4c22-80a9-582e5fe3465a" "74512031-b546-486c-99a4-3b2a99f7c415" "69ce50ee-ec05-48fb-bd69-0dac46ed343c" ...
##  $ submitter_id         : chr "TCGA-3A-A9IJ"
##  $ state                : NULL
##  $ aliquot_ids          : chr [1:12] "02ddf7fb-22dc-4dcd-bb9c-bfcab024a730" "c31e1d22-572b-46a0-8871-fd5ad89afbbc" "7756dde4-1712-4cec-bbe0-77d736db7ddd" "7c45683b-8b17-450e-8bb9-4f035882bc0d" ...
##  $ slide_ids            : chr "fba3ed3d-2167-4207-8117-6ca64c0ed4a1"
##  $ submitter_sample_ids : chr [1:2] "TCGA-3A-A9IJ-10A" "TCGA-3A-A9IJ-01A"</code></pre>
</div>
<div id="annotations" class="section level3">
<h3 class="hasAnchor">
<html><body><a href="#annotations" class="anchor"> </a></body></html>Annotations</h3>
<p>Annotations are notes added to individual cases, samples or files. They can be searched and filtered in the same way as other entities. See <a href="https://docs.gdc.cancer.gov/Data_Portal/Users_Guide/Annotations/">here</a> for details from Data Portal about annotations.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">a =<span class="st"> </span><span class="kw"><a href="../reference/query.html">annotations</a></span>()
annots_df =<span class="st"> </span><span class="kw">do.call</span>(rbind, a)
<span class="kw">datatable</span>(annots_df[,<span class="kw">c</span>(<span class="st">'category'</span>,<span class="st">'classification'</span>,<span class="st">'notes'</span>,<span class="st">'entity_type'</span>)],<span class="dt">rownames=</span><span class="ot">FALSE</span>)</code></pre></div>
<div id="htmlwidget-0b33459ce0afcfb66074" style="width:100%;height:auto;" class="datatables html-widget"></div>
<script type="application/json" data-for="htmlwidget-0b33459ce0afcfb66074">{"x":{"filter":"none","data":[["Neoadjuvant therapy","Center QC failed","Center QC failed","Alternate sample pipeline","History of acceptable prior treatment related to a prior/other malignancy","Synchronous malignancy","Prior malignancy","Prior malignancy","Center QC failed","Item flagged DNU"],["Notification","CenterNotification","CenterNotification","Notification","Notification","Notification","Notification","Notification","CenterNotification","CenterNotification"],["Interferon was given less than 90 days prior to cancer procurement, which is exclusionary. (per Kristen Leraas/BCR TCGA-2222)","RNA-seq:LOW 5/3 COVERAGE RATIO;HIGH MITOCHONDRIAL CONTENT","RNA-seq:INSUFFICIENT INPUT MATERIAL,LOW SEQUENCE YIELD/DIVERSITY;LOW 5/3 COVERAGE RATIO","Biospecimens from this case belong to Batch 25 and were processed into analyte outside of the normal TCGA standardized laboratory pipeline.  Users should note that the resulting characterization data may exhibit features (e.g. batch effects) due to this difference.","Prior malignancy SCC of the tongue treated with locoregional radiation.","Pt with synchronous thymoma.","Prior malignancy breast cancer treated with radiation","Prior malignancy of breast ductal and lobular carcinoma","RNA-seq:LOW 5/3 COVERAGE RATIO","The aliquot failed Broad pipeline QC and not all files are suitable for use. Consult the SDRF file to determine which files are usable."],["case","aliquot","aliquot","case","case","case","case","case","aliquot","aliquot"]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th>category\u003c/th>\n      <th>classification\u003c/th>\n      <th>notes\u003c/th>\n      <th>entity_type\u003c/th>\n    \u003c/tr>\n  \u003c/thead>\n\u003c/table>","options":{"order":[],"autoWidth":false,"orderClasses":false}},"evals":[],"jsHooks":[]}</script>
</div>
<div id="advanced-querying-with-filters" class="section level3">
<h3 class="hasAnchor">
<html><body><a href="#advanced-querying-with-filters" class="anchor"> </a></body></html>Advanced querying with filters</h3>
<p>The GDC supports <a href="https://docs.gdc.cancer.gov/Data_Portal/Users_Guide/Advanced_Search/">advanced queries</a> based on filters and facets. The GenomicDataCommons package allows writing queries using standard R syntax, simplifying the creation of complicated filtering patterns. Examples follow:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># youngcases = cases(filters=filters(diagnoses.age_at_diagnosis &lt;= 10*365,endpoint='cases'))</span></code></pre></div>
</div>
</div>
<div id="downloading-data" class="section level2 tabset tabset-fade">
<h2 class="hasAnchor">
<html><body><a href="#downloading-data" class="anchor"> </a></body></html>Downloading Data</h2>
<p>Downloading files from the GDC starts starts with identifying files of interest and collecting the file IDs which the GDC uses as keys for the files. Then, the transfer can be done using can be done using directly with R-based tooling for small transfers of smaller single files or multiple small files. For larger files such as any genomics data files, using a specialized data transfer (<code>gdc-transfer</code>) tool is likely to be more efficient and offers the potential for restarting transfers. The two sections below deal with these two approaches.</p>
<p>BAM file slicing is an operation specific to querying pieces of BAM files from the GDC.</p>
<div id="single-or-multiple-small-file-downloads" class="section level3">
<h3 class="hasAnchor">
<html><body><a href="#single-or-multiple-small-file-downloads" class="anchor"> </a></body></html>Single or multiple small file downloads</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">a =<span class="st"> </span><span class="kw"><a href="../reference/query.html">files</a></span>(<span class="dt">filters=</span><span class="kw"><a href="../reference/make_filter.html">make_filter</a></span>(<span class="st">'data_format'</span>==<span class="st">'MAF'</span> &amp;<span class="st"> 'access'</span>==<span class="st">'open'</span>))
fpaths =<span class="st"> </span><span class="kw"><a href="../reference/gdcdata.html">gdcdata</a></span>(<span class="dt">uuids=</span><span class="kw">names</span>(a),<span class="dt">progress=</span><span class="ot">FALSE</span>)
<span class="kw">basename</span>(fpaths)</code></pre></div>
<pre><code>##  [1] "TCGA.DLBC.varscan.c2a80453-d1e5-44a5-8dd1-c1d473227686.somatic.maf.gz"      
##  [2] "TCGA.READ.varscan.4683cb8b-94d6-4b98-8886-c9e85542d060.somatic.maf.gz"      
##  [3] "TCGA.KIRP.somaticsniper.d01afe22-9edf-4b90-b812-ecb7bb021240.somatic.maf.gz"
##  [4] "TCGA.KIRC.somaticsniper.c2ff40fb-8452-4791-a19d-6861bf89a619.somatic.maf.gz"
##  [5] "TCGA.THYM.mutect.7c714e03-b52c-4a82-aac5-b09ee5b41131.somatic.maf.gz"       
##  [6] "TCGA.TGCT.mutect.0054476f-0a44-4eba-9f08-e06492d41747.somatic.maf.gz"       
##  [7] "TCGA.UCEC.varscan.0517d207-45d4-46fa-877d-e7aa7d60e63d.somatic.maf.gz"      
##  [8] "TCGA.KIRP.muse.18188602-bc86-46ff-aff7-baff03114282.somatic.maf.gz"         
##  [9] "TCGA.ACC.varscan.f258cc30-bfee-4e3b-bc55-a1105235ffea.somatic.maf.gz"       
## [10] "TCGA.READ.mutect.11da8452-8316-4b29-90fa-12cdaa18c1b7.somatic.maf.gz"</code></pre>
</div>
<div id="large-or-many-files-with-gdc-download-tool" class="section level3">
<h3 class="hasAnchor">
<html><body><a href="#large-or-many-files-with-gdc-download-tool" class="anchor"> </a></body></html>Large or many files with GDC download tool</h3>
<p>For many files or large files, the [GDC Download Tool][] can be more efficient and allows for download restarts if necessary.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">manifest_path =<span class="st"> </span><span class="kw"><a href="../reference/manifest.html">manifest</a></span>(<span class="dt">uuids =</span> <span class="kw">names</span>(a))
manifest_path</code></pre></div>
<pre><code>## [1] "/var/folders/21/b_rp6qyj1_b1j5cp8qby0tnr0000gn/T//RtmpEYWgmv/file97ee64c68202/gdc_manifest_20170123_235325.txt"</code></pre>
<p>Use the [GDC Download Tool][] for your platform and to read and then download files in the <code>manifest_path</code> from above.</p>
</div>
</div>
<div id="sessioninfo" class="section level2">
<h2 class="hasAnchor">
<html><body><a href="#sessioninfo" class="anchor"> </a></body></html>SessionInfo</h2>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">sessionInfo</span>()</code></pre></div>
<pre><code>## R Under development (unstable) (2016-10-26 r71594)
## Platform: x86_64-apple-darwin13.4.0 (64-bit)
## Running under: macOS Sierra 10.12.2
## 
## locale:
## [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8
## 
## attached base packages:
## [1] stats     graphics  grDevices utils     datasets  methods   base     
## 
## other attached packages:
## [1] GenomicDataCommons_0.3.0 jsonlite_1.2            
## [3] DT_0.2                   knitr_1.15.1            
## 
## loaded via a namespace (and not attached):
##  [1] Rcpp_0.12.8.4        roxygen2_5.0.1       xml2_1.1.0          
##  [4] magrittr_1.5         devtools_1.12.0.9000 pkgload_0.0.0.9000  
##  [7] R6_2.2.0             stringr_1.1.0        httr_1.2.1          
## [10] tools_3.4.0          pkgbuild_0.0.0.9000  withr_1.0.2         
## [13] htmltools_0.3.5      yaml_2.1.14          rprojroot_1.1       
## [16] digest_0.6.11        crayon_1.3.2         htmlwidgets_0.8     
## [19] codetools_0.2-15     testthat_1.0.2.9000  curl_2.3            
## [22] memoise_1.0.0        evaluate_0.10        rmarkdown_1.3       
## [25] stringi_1.1.2        compiler_3.4.0       BiocInstaller_1.25.2
## [28] backports_1.0.4</code></pre>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li>
<a href="#the-api">The API</a><ul class="nav nav-pills nav-stacked">
<li><a href="#querying-metadata">Querying Metadata</a></li>
      <li><a href="#downloading-data">Downloading Data</a></li>
      <li><a href="#sessioninfo">SessionInfo</a></li>
      </ul>
</li>
      </ul>
</div>
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by Martin Morgan, Davis Sean.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://hadley.github.io/pkgdown/">pkgdown</a>.</p>
</div>

      </footer>
</div>

  </body>
</html>
