<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Introduction to Accessing the NCI Genomic Data Commons in Bioconductor &bull; GenomicDataCommons</title><!-- jquery --><script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous"><script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script><!-- Font Awesome icons --><link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous"><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../jquery.sticky-kit.min.js"></script><script src="../pkgdown.js"></script><!-- mathjax --><script src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <div class="container template-vignette">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="../index.html">GenomicDataCommons</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../articles/index.html">Articles</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"></ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header><script src="overview_files/htmlwidgets-0.8/htmlwidgets.js"></script><script src="overview_files/jquery-1.12.4/jquery.min.js"></script><script src="overview_files/datatables-binding-0.2/datatables.js"></script><link href="overview_files/dt-core-1.10.12/css/jquery.dataTables.min.css" rel="stylesheet"><link href="overview_files/dt-core-1.10.12/css/jquery.dataTables.extra.css" rel="stylesheet"><script src="overview_files/dt-core-1.10.12/js/jquery.dataTables.min.js"></script><div class="row">
  <div class="col-md-9">
    <div class="page-header toc-ignore">
      <h1>Introduction to Accessing the NCI Genomic Data Commons in Bioconductor</h1>
                        <h4 class="author">Sean Davis &amp; Martin Morgan</h4>
            
            <h4 class="date">2016-12-18</h4>
          </div>

    
    
<div class="contents">
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(GenomicDataCommons)
<span class="kw">library</span>(DT)</code></pre></div>
<div id="background" class="section level1">
<h1 class="hasAnchor"><html><body><a href="#background" class="anchor"> </a></body></html>Background</h1>
<p>From the <a href="https://gdc.nci.nih.gov/about-gdc">Genomic Data Commons (GDC) website</a>:</p>
<p>The National Cancer Institute&rsquo;s (NCI&rsquo;s) Genomic Data Commons (GDC) is a data sharing platform that promotes precision medicine in oncology. It is not just a database or a tool; it is an expandable knowledge network supporting the import and standardization of genomic and clinical data from cancer research programs.</p>
<p>The GDC contains NCI-generated data from some of the largest and most comprehensive cancer genomic datasets, including The Cancer Genome Atlas (TCGA) and Therapeutically Applicable Research to Generate Effective Therapies (TARGET). For the first time, these datasets have been harmonized using a common set of bioinformatics pipelines, so that the data can be directly compared.</p>
<p>As a growing knowledge system for cancer, the GDC also enables researchers to submit data, and harmonizes these data for import into the GDC. As more researchers add clinical and genomic data to the GDC, it will become an even more powerful tool for making discoveries about the molecular basis of cancer that may lead to better care for patients.</p>
<div id="data-model" class="section level2">
<h2 class="hasAnchor"><html><body><a href="#data-model" class="anchor"> </a></body></html>Data Model</h2>
<p>The <a href="https://gdc.cancer.gov/developers/gdc-data-model/gdc-data-model-components">data model for the GDC is complex</a>, but it worth a quick overview. The data model is encoded as a so-called property graph. Nodes represent entities such as Projects, Cases, Diagnoses, Files (various kinds), and Annotations. The relationships between these entities are maintained as edges. Both nodes and edges may have Properties that supply instance details. The GDC API exposes these nodes and edges in a somewhat simplified set of <a href="https://en.wikipedia.org/wiki/Representational_state_transfer">RESTful</a> endpoints.</p>
</div>
<div id="full-feature-set-of-the-gdc-api" class="section level2">
<h2 class="hasAnchor"><html><body><a href="#full-feature-set-of-the-gdc-api" class="anchor"> </a></body></html>Full Feature Set of the GDC API</h2>
<p>HTTP interface that uses JSON as the primary data exchange format</p>
<ul><li>Programmatic access to functionality provided by GDC Data and Submission portals, via projects, cases, files, annotations, data, slicing, status, and submission endpoints</li>
<li>Programmatic access to GDC Legacy Archive via legacy endpoint</li>
<li>Token-based authentication for secure access to controlled data and to submission functionality</li>
<li>RESTful search that supports simple and complex queries via filters, fields, and facets parameters, and project, files, cases, and annotations endpoints.</li>
<li>Search results can be sorted using sort parameter, paginated using size and from parameters, and output in JSON, TSV, and XML using format and pretty parameters.</li>
<li>_mapping endpoint enables user discovery of fields available for data search and retrieval operations</li>
<li>Support for downloading of individual files and of archives containing multiple files</li>
<li>[NOT Implemented Here] Generation of download and upload manifests for use with the GDC Data Transfer Tool</li>
<li>BAM slicing functionality for downloading part(s) of a BAM file specified using chromosomal coordinates or HGNC gene names</li>
<li>[NOT Implemented Here] Transactional submission system that links individual data elements according to a graph-based GDC Data Model</li>
<li>Two data entity identifiers: UUIDs, which are consistent across GDC, and Submitter IDs, for compatibility with submitters&rsquo; tracking systems</li>
</ul><p>Here, we will be describing how to access the main API endpoints:</p>
<ul><li><a href="https://docs.gdc.cancer.gov/Data_Portal/Users_Guide/Projects/">Projects</a></li>
<li><a href="https://docs.gdc.cancer.gov/Data_Portal/Users_Guide/Cases_and_Files/">Files</a></li>
<li><a href="https://docs.gdc.cancer.gov/Data_Portal/Users_Guide/Cases_and_Files/">Cases</a></li>
<li><a href="https://docs.gdc.cancer.gov/Data_Portal/Users_Guide/Annotations/">Annotations</a></li>
<li><a href="https://docs.gdc.cancer.gov/Data_Portal/Users_Guide/Cases_and_Files/#bam-slicing">BAM Slicing</a> - Specific to dealing with BAM files</li>
<li>_mappings - For obtaining metadata about fields associated with the various endpoints</li>
</ul><p>The following sections address each in turn, including examples of usage.</p>
</div>
</div>
<div id="the-basic-api-endpoints" class="section level1">
<h1 class="hasAnchor"><html><body><a href="#the-basic-api-endpoints" class="anchor"> </a></body></html>The basic API endpoints</h1>
<p>The Projects, Files, Cases, and Annotations endpoints are are all accessed using functions of the same name (all lowercase). The default behavior of each is to query the GDC and return the first 10 records with a default set of fields. In this section, the results are simply converted to a tabular form, but the data are not always tabular and may have one-to-many relationships, so this represents a simplification. Accessing the data more directly will be covered in later sections.</p>
<div id="projects" class="section level2">
<h2 class="hasAnchor"><html><body><a href="#projects" class="anchor"> </a></body></html>Projects</h2>
<p>To get a sense of what is available in the default Project entities, we simply call the <code><a href="../reference/project_fields.html">projects()</a></code> function without any arguments. By default, the first 10 projects are returned. The returned value for projects is a <code>data.frame</code> since the Project entity is tabular.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">a =<span class="st"> </span><span class="kw"><a href="../reference/project_fields.html">projects</a></span>()</code></pre></div>
<pre><code>## GET request uri:
## https://gdc-api.nci.nih.gov/projects?format=XML&amp;pretty=FALSE&amp;fields=dbgap_accession_number,disease_type,released,state,primary_site,project_id,name&amp;size=10&amp;from=1</code></pre>
<pre><code>## No encoding supplied: defaulting to UTF-8.</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">datatable</span>(a)</code></pre></div>
<div id="htmlwidget-16882469dcde2d404c95" style="width:100%;height:auto;" class="datatables html-widget"></div>
<script type="application/json" data-for="htmlwidget-16882469dcde2d404c95">{"x":{"filter":"none","data":[["1","2","3","4","5","6","7","8","9","10"],["","","phs000467","","","","","","","phs000468"],["Sarcoma","Pancreatic Adenocarcinoma","Neuroblastoma","Rectum Adenocarcinoma","Mesothelioma","Glioblastoma Multiforme","Kidney Renal Papillary Cell Carcinoma","Adrenocortical Carcinoma","Cervical Squamous Cell Carcinoma and Endocervical Adenocarcinoma","Osteosarcoma"],["true","true","true","true","true","true","true","true","true","true"],["legacy","legacy","legacy","legacy","legacy","legacy","legacy","legacy","legacy","legacy"],["Soft Tissue","Pancreas","Nervous System","Colorectal","Pleura","Brain","Kidney","Adrenal Gland","Cervix","Bone"],["TCGA-SARC","TCGA-PAAD","TARGET-NBL","TCGA-READ","TCGA-MESO","TCGA-GBM","TCGA-KIRP","TCGA-ACC","TCGA-CESC","TARGET-OS"],["Sarcoma","Pancreatic Adenocarcinoma","Neuroblastoma","Rectum Adenocarcinoma","Mesothelioma","Glioblastoma Multiforme","Kidney Renal Papillary Cell Carcinoma","Adrenocortical Carcinoma","Cervical Squamous Cell Carcinoma and Endocervical Adenocarcinoma","Osteosarcoma"]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th> \u003c/th>\n      <th>dbgap_accession_number\u003c/th>\n      <th>disease_type\u003c/th>\n      <th>released\u003c/th>\n      <th>state\u003c/th>\n      <th>primary_site\u003c/th>\n      <th>project_id\u003c/th>\n      <th>name\u003c/th>\n    \u003c/tr>\n  \u003c/thead>\n\u003c/table>","options":{"order":[],"autoWidth":false,"orderClasses":false,"columnDefs":[{"orderable":false,"targets":0}]}},"evals":[],"jsHooks":[]}</script></div>
<div id="files" class="section level2">
<h2 class="hasAnchor"><html><body><a href="#files" class="anchor"> </a></body></html>Files</h2>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">fields &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">"data_type"</span>, <span class="st">"updated_datetime"</span>, <span class="st">"created_datetime"</span>,
    <span class="st">"file_name"</span>, <span class="st">"md5sum"</span>, <span class="st">"data_format"</span>, <span class="st">"acl"</span>, <span class="st">"access"</span>, <span class="st">"state"</span>,
    <span class="st">"data_category"</span>, <span class="st">"file_size"</span>, <span class="st">"submitter_id"</span>, <span class="st">"type"</span>)
a =<span class="st"> </span><span class="kw"><a href="../reference/file_fields.html">files</a></span>(<span class="dt">fields=</span>fields)</code></pre></div>
<pre><code>## GET request uri:
## https://gdc-api.nci.nih.gov/files?format=JSON&amp;pretty=FALSE&amp;fields=file_id,data_type,updated_datetime,created_datetime,file_name,md5sum,data_format,acl,access,state,data_category,file_size,submitter_id,type&amp;size=10&amp;from=1</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">datatable</span>(<span class="kw">do.call</span>(rbind, a))</code></pre></div>
<div id="htmlwidget-226c18a0138ff1ccc55a" style="width:100%;height:auto;" class="datatables html-widget"></div>
<script type="application/json" data-for="htmlwidget-226c18a0138ff1ccc55a">{"x":{"filter":"none","data":[["8bfeb962-e2b3-463a-9fbb-f5c6296f3ec2","fd8a5011-5dfa-4a9d-ae21-f638b9616153","49c24b40-f086-42bf-b591-253910f63def","a43c4463-dc31-450e-9a12-64cf9dedc96d","a7fd6aae-6af5-490f-b65e-97c7bfcf44bf","164511a9-2f56-49e0-b5cf-9c4be32f8fc7","ce8d8ba3-fdd7-4b3c-9bba-20c3f578d277","796ea613-f0da-4592-a29e-74edf27a442a","4dd5ba61-80ac-440f-aced-ce0d3d468f6b","61cb79e3-ec5f-4211-b41d-64ff63b30d0e"],["Annotated Somatic Mutation","Methylation Beta Value","Clinical Supplement","Masked Copy Number Segment","Aligned Reads","Aligned Reads","Methylation Beta Value","Raw Simple Somatic Mutation","Annotated Somatic Mutation","Aligned Reads"],["2016-09-29T21:15:55.100009-05:00","2016-10-27T21:58:12.297090-05:00","2016-09-14T11:25:51.615985-05:00","2016-09-29T21:54:58.373972-05:00","2016-09-29T22:01:04.154586-05:00","2016-09-29T18:06:42.655488-05:00","2016-10-27T21:58:12.297090-05:00","2016-09-29T20:46:17.811696-05:00","2016-09-29T19:35:19.556067-05:00","2016-09-29T20:07:45.862432-05:00"],["2016-05-04T16:05:36.917600-05:00","2016-10-27T21:58:12.297090-05:00","2016-05-04T09:11:20.201320-05:00","2016-05-03T03:29:52.126663-05:00","2016-05-28T18:38:39.666793-05:00","2016-05-03T13:43:08.390772-05:00","2016-10-27T21:58:12.297090-05:00","2016-05-04T15:21:07.352135-05:00","2016-05-29T11:00:12.508960-05:00","2016-05-29T23:50:17.072708-05:00"],["8bfeb962-e2b3-463a-9fbb-f5c6296f3ec2.vep.reheader.vcf.gz","jhu-usc.edu_THCA.HumanMethylation450.3.lvl-3.TCGA-EM-A1YD-11A-11D-A14Z-05.gdc_hg38.txt","nationwidechildrens.org_clinical.TCGA-QT-A5XO.xml","SNOWS_p_TCGA_Batch_309_310_NSP_GenomeWideSNP_6_C01_1362120.nocnv_grch38.seg.txt","f49429e8-6f2b-4546-af98-bbcc17819f0c_gdc_realn_rehead.bam","C529.TCGA-HC-7737-11A-02D-2114-08.1_gdc_realn.bam","jhu-usc.edu_LAML.HumanMethylation450.2.lvl-3.TCGA-AB-2955-03A-01D-0742-05.gdc_hg38.txt","796ea613-f0da-4592-a29e-74edf27a442a.vcf.gz","4dd5ba61-80ac-440f-aced-ce0d3d468f6b.vep.reheader.vcf.gz","C495.TCGA-QK-A6V9-01A-11D-A34J-08.2_gdc_realn.bam"],["ace85792006140a9b4895976ecba725e","36ffbc7fea6400383691afde8b19f2ee","39704471366afcb874d8cee790c68a67","ab0dcb6728eeb0caca87bb8d56d504eb","8f3cdc3bf16f0d6de5cea19958ecdedc","ac144464dce61d05a58e7d83530c2a18","c0aa7cd2700dab2298cf25d4da581151","5890195a2251f8102b0cd34b8ad08b33","9cc457a70e3d4c8ecd5cfe936fe7a263","7159005e5eef731cd5210205cd6026a9"],["VCF","TXT","BCR XML","TXT","BAM","BAM","TXT","VCF","VCF","BAM"],["phs000178","open","open","open","phs000178","phs000178","open","phs000178","phs000178","phs000178"],["controlled","open","open","open","controlled","controlled","open","controlled","controlled","controlled"],["submitted","submitted","live","submitted","submitted","submitted","submitted","submitted","submitted","submitted"],["Simple Nucleotide Variation","DNA Methylation","Clinical","Copy Number Variation","Raw Sequencing Data","Raw Sequencing Data","DNA Methylation","Simple Nucleotide Variation","Simple Nucleotide Variation","Raw Sequencing Data"],[378676,141301769,20982,4536,2061824371,12036462076,141280622,160084,302033,17038419572],["TCGA-E3-A3E3-01A-11D-A20C-08_TCGA-E3-A3E3-10A-01D-A20C-08_somaticsniper_annotated","77eb2c26-b386-4254-94c5-4fcf83993221-beta-value",null,"a8545cfc-fbaf-4b3d-a539-411818df3934_nocnv","f49429e8-6f2b-4546-af98-bbcc17819f0c","606ca8e2-3442-4b78-899c-9b27e292b075","185c04ce-6b51-416a-92fc-d3ecfdce3d6e-beta-value","TCGA-44-2656-01B-06D-A271-08_TCGA-44-2656-10A-01D-A271-08_somaticsniper","TCGA-AR-A5QP-01A-11D-A28B-09_TCGA-AR-A5QP-10A-01D-A28E-09_somaticsniper_annotated","c47e6c13-8664-4589-b89b-893518621312"],["annotated_somatic_mutation","methylation_beta_value","clinical_supplement","copy_number_segment","aligned_reads","aligned_reads","methylation_beta_value","simple_somatic_mutation","annotated_somatic_mutation","aligned_reads"]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th> \u003c/th>\n      <th>data_type\u003c/th>\n      <th>updated_datetime\u003c/th>\n      <th>created_datetime\u003c/th>\n      <th>file_name\u003c/th>\n      <th>md5sum\u003c/th>\n      <th>data_format\u003c/th>\n      <th>acl\u003c/th>\n      <th>access\u003c/th>\n      <th>state\u003c/th>\n      <th>data_category\u003c/th>\n      <th>file_size\u003c/th>\n      <th>submitter_id\u003c/th>\n      <th>type\u003c/th>\n    \u003c/tr>\n  \u003c/thead>\n\u003c/table>","options":{"order":[],"autoWidth":false,"orderClasses":false,"columnDefs":[{"orderable":false,"targets":0}]}},"evals":[],"jsHooks":[]}</script></div>
<div id="cases" class="section level2">
<h2 class="hasAnchor"><html><body><a href="#cases" class="anchor"> </a></body></html>Cases</h2>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">a =<span class="st"> </span><span class="kw"><a href="../reference/case_fields.html">cases</a></span>()</code></pre></div>
<pre><code>## GET request uri:
## https://gdc-api.nci.nih.gov/cases?format=JSON&amp;pretty=FALSE&amp;fields=sample_ids,portion_ids,submitter_portion_ids,created_datetime,submitter_aliquot_ids,updated_datetime,submitter_analyte_ids,analyte_ids,submitter_id,case_id,state,aliquot_ids,slide_ids,submitter_sample_ids&amp;size=10&amp;from=1</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">a             <span class="co"># 10 cases by default</span></code></pre></div>
<pre><code>## class: cases_list
## cases: 10
## names:
##     5f468c25-73c3-49d5-ba12-50653dfe2c41,
##     52c35691-efca-42f5-a03e-02c9f872275c,
##     5f99343b-e8ea-4b41-b36d-cd5c8fd918d0, ...,
##     643ac1e9-3b8d-5cfa-ab52-18de4bc8d3d3,
##     fcf64b55-5c4f-4d82-ac6a-4713d01143cb</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">lengths</span>(a)    <span class="co"># cases have different elements</span></code></pre></div>
<pre><code>## 5f468c25-73c3-49d5-ba12-50653dfe2c41 52c35691-efca-42f5-a03e-02c9f872275c 
##                                   13                                   13 
## 5f99343b-e8ea-4b41-b36d-cd5c8fd918d0 fafd6f5b-1d76-4537-bd1c-e0bd7b4e2166 
##                                   13                                   13 
## 183dd089-e932-4be2-b252-0e8572e7da4e 9283709d-d019-4291-a097-611a5c05244b 
##                                   13                                   13 
## 89c72f0e-e724-4826-bb9c-d7524d1b1258 929837f9-297a-47c0-ac9a-a142eef620d3 
##                                   13                                   13 
## 643ac1e9-3b8d-5cfa-ab52-18de4bc8d3d3 fcf64b55-5c4f-4d82-ac6a-4713d01143cb 
##                                    6                                   13</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">a[<span class="dv">1</span>]          <span class="co"># summary</span></code></pre></div>
<pre><code>## class: cases_list
## cases: 1
## names:
##     5f468c25-73c3-49d5-ba12-50653dfe2c41
## fields:
##     sample_ids(2)
##     portion_ids(2)
##     submitter_portion_ids(2)
##     created_datetime(0)
##     submitter_aliquot_ids(12)
##     updated_datetime(1)
##     submitter_analyte_ids(5)
##     analyte_ids(5)
##     submitter_id(1)
##     state(0)
##     aliquot_ids(12)
##     slide_ids(1)
##     submitter_sample_ids(2)</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">head</span>(a[[<span class="dv">1</span>]])  <span class="co"># values</span></code></pre></div>
<pre><code>## $sample_ids
## [1] "5d809b44-5c5e-4942-ac8e-a5650e1e1ccc"
## [2] "f80e9141-3634-49cd-8c85-610c7f644d83"
## 
## $portion_ids
## [1] "08d8cb7b-1725-4f3e-8e16-989a232de883"
## [2] "35e98f3f-fa01-4b37-9892-7292aa908b6f"
## 
## $submitter_portion_ids
## [1] "TCGA-3A-A9IJ-01A-11" "TCGA-3A-A9IJ-10A-01"
## 
## $created_datetime
## NULL
## 
## $submitter_aliquot_ids
##  [1] "TCGA-3A-A9IJ-01A-11D-A398-05" "TCGA-3A-A9IJ-01A-11R-A39J-13"
##  [3] "TCGA-3A-A9IJ-01A-11D-A395-26" "TCGA-3A-A9IJ-01A-11R-A39D-07"
##  [5] "TCGA-3A-A9IJ-10A-01D-A395-26" "TCGA-3A-A9IJ-10A-01D-A399-01"
##  [7] "TCGA-3A-A9IJ-10A-01D-A39A-08" "TCGA-3A-A9IJ-01A-01D-YYYY-23"
##  [9] "TCGA-3A-A9IJ-01A-11D-A396-01" "TCGA-3A-A9IJ-01A-11W-A450-08"
## [11] "TCGA-3A-A9IJ-01A-11D-A397-08" "TCGA-3A-A9IJ-10A-01W-A451-08"
## 
## $updated_datetime
## [1] "2016-09-08T10:25:08.392801-05:00"</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">str</span>(a[[<span class="dv">1</span>]])</code></pre></div>
<pre><code>## List of 13
##  $ sample_ids           : chr [1:2] "5d809b44-5c5e-4942-ac8e-a5650e1e1ccc" "f80e9141-3634-49cd-8c85-610c7f644d83"
##  $ portion_ids          : chr [1:2] "08d8cb7b-1725-4f3e-8e16-989a232de883" "35e98f3f-fa01-4b37-9892-7292aa908b6f"
##  $ submitter_portion_ids: chr [1:2] "TCGA-3A-A9IJ-01A-11" "TCGA-3A-A9IJ-10A-01"
##  $ created_datetime     : NULL
##  $ submitter_aliquot_ids: chr [1:12] "TCGA-3A-A9IJ-01A-11D-A398-05" "TCGA-3A-A9IJ-01A-11R-A39J-13" "TCGA-3A-A9IJ-01A-11D-A395-26" "TCGA-3A-A9IJ-01A-11R-A39D-07" ...
##  $ updated_datetime     : chr "2016-09-08T10:25:08.392801-05:00"
##  $ submitter_analyte_ids: chr [1:5] "TCGA-3A-A9IJ-01A-11R" "TCGA-3A-A9IJ-01A-11D" "TCGA-3A-A9IJ-10A-01D" "TCGA-3A-A9IJ-10A-01W" ...
##  $ analyte_ids          : chr [1:5] "57518237-1d1e-4c17-967f-3de1ddff7e50" "e995c697-d0de-4c22-80a9-582e5fe3465a" "74512031-b546-486c-99a4-3b2a99f7c415" "69ce50ee-ec05-48fb-bd69-0dac46ed343c" ...
##  $ submitter_id         : chr "TCGA-3A-A9IJ"
##  $ state                : NULL
##  $ aliquot_ids          : chr [1:12] "02ddf7fb-22dc-4dcd-bb9c-bfcab024a730" "c31e1d22-572b-46a0-8871-fd5ad89afbbc" "7756dde4-1712-4cec-bbe0-77d736db7ddd" "7c45683b-8b17-450e-8bb9-4f035882bc0d" ...
##  $ slide_ids            : chr "fba3ed3d-2167-4207-8117-6ca64c0ed4a1"
##  $ submitter_sample_ids : chr [1:2] "TCGA-3A-A9IJ-10A" "TCGA-3A-A9IJ-01A"</code></pre>
</div>
<div id="annotations" class="section level2">
<h2 class="hasAnchor"><html><body><a href="#annotations" class="anchor"> </a></body></html>Annotations</h2>
<p>Annotations are notes added to individual cases, samples or files. They can be searched and filtered in the same way as other entities. See <a href="https://docs.gdc.cancer.gov/Data_Portal/Users_Guide/Annotations/">here</a> for details from Data Portal about annotations.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">a =<span class="st"> </span><span class="kw"><a href="../reference/annotation_fields.html">annotations</a></span>()</code></pre></div>
<pre><code>## GET request uri:
## https://gdc-api.nci.nih.gov/annotations?format=JSON&amp;pretty=FALSE&amp;fields=status,category,entity_id,classification,updated_datetime,created_datetime,annotation_id,notes,entity_type,submitter_id,case_id,entity_submitter_id,case_submitter_id&amp;size=10&amp;from=1</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">datatable</span>(<span class="kw">do.call</span>(rbind, a))</code></pre></div>
<div id="htmlwidget-dc8be52a5d12430deb20" style="width:100%;height:auto;" class="datatables html-widget"></div>
<script type="application/json" data-for="htmlwidget-dc8be52a5d12430deb20">{"x":{"filter":"none","data":[["02aa12cb-7108-5710-9d55-6b10555b8e39","aebc7105-ccf9-5d51-9ef0-f47b624fc5a3","5cf05f41-ce70-58a3-8ecb-6bfaf6264437","d5a0cebe-e917-5c99-9178-6c41635b43dd","7264f17f-6b6d-5d17-9ce4-648cf77abfdf","c0642913-2e40-5ded-8eed-ab2dc0db5ae5","bbc5016f-337d-5e10-9e15-ca81d82e0d2d","b5ede1f3-b042-5944-8188-ad73d5b53de2","459ab6a6-849e-5da2-ba03-aa64af6591b1","ae9291b5-7960-5322-903c-5a1cc12b71dd"],["Approved","Approved","Approved","Approved","Approved","Approved","Approved","Approved","Approved","Approved"],["Neoadjuvant therapy","Center QC failed","Center QC failed","Alternate sample pipeline","History of acceptable prior treatment related to a prior/other malignancy","Synchronous malignancy","Prior malignancy","Prior malignancy","Center QC failed","Item flagged DNU"],["b85b14d7-3b5a-4800-af12-622ec03b9fe5","520e1d91-a556-4443-8c08-feec29c0c4fe","733f0607-6c6b-4385-9868-fa6f155a9a2e","b8492e1a-aa02-4585-a5be-75f9833c8c3f","49f975ea-a1cf-4a9b-bf13-030c68cc99e4","c0fdb152-25d2-404b-bec7-a43ece381f5b","c16d9f69-26f9-4ec7-b2d5-ff9f7dacfe0f","81d11171-5d9e-4950-98f8-b0ab0f2d7908","5b759ce4-2632-4d85-89fe-30a85069e9a8","cb0b38e9-ae39-4f69-b158-a2b78d13177e"],["Notification","CenterNotification","CenterNotification","Notification","Notification","Notification","Notification","Notification","CenterNotification","CenterNotification"],["2016-05-01T15:00:21.638875-05:00","2016-05-01T15:00:21.638875-05:00","2016-05-01T15:00:21.638875-05:00","2016-05-01T15:00:21.638875-05:00","2016-05-01T15:00:21.638875-05:00","2016-05-01T15:00:21.638875-05:00","2016-05-01T15:00:21.638875-05:00","2016-05-01T15:00:21.638875-05:00","2016-05-01T15:00:21.638875-05:00","2016-05-01T15:00:21.638875-05:00"],["2014-08-14T14:25:49-04:00","2012-07-20T21:54:03-04:00","2012-07-20T21:52:24-04:00","2012-11-13T12:43:36-05:00","2012-11-10T05:24:04-05:00","2012-11-10T05:25:39-05:00","2012-10-31T14:33:07-04:00","2011-09-18T20:47:10-04:00","2012-07-20T21:52:05-04:00","2015-09-28T13:41:01-04:00"],["Interferon was given less than 90 days prior to cancer procurement, which is exclusionary. (per Kristen Leraas/BCR TCGA-2222)","RNA-seq:LOW 5/3 COVERAGE RATIO;HIGH MITOCHONDRIAL CONTENT","RNA-seq:INSUFFICIENT INPUT MATERIAL,LOW SEQUENCE YIELD/DIVERSITY;LOW 5/3 COVERAGE RATIO","Biospecimens from this case belong to Batch 25 and were processed into analyte outside of the normal TCGA standardized laboratory pipeline.  Users should note that the resulting characterization data may exhibit features (e.g. batch effects) due to this difference.","Prior malignancy SCC of the tongue treated with locoregional radiation.","Pt with synchronous thymoma.","Prior malignancy breast cancer treated with radiation","Prior malignancy of breast ductal and lobular carcinoma","RNA-seq:LOW 5/3 COVERAGE RATIO","The aliquot failed Broad pipeline QC and not all files are suitable for use. Consult the SDRF file to determine which files are usable."],["case","aliquot","aliquot","case","case","case","case","case","aliquot","aliquot"],["21869","8811","8764","12330","11936","11971","11839","3396","8755","29302"],["b85b14d7-3b5a-4800-af12-622ec03b9fe5","f8a5547f-f1bc-4c01-8b68-25b5ee0caeab","3e8a51bf-7e1f-4eab-af83-3c60d04db1bf","b8492e1a-aa02-4585-a5be-75f9833c8c3f","49f975ea-a1cf-4a9b-bf13-030c68cc99e4","c0fdb152-25d2-404b-bec7-a43ece381f5b","c16d9f69-26f9-4ec7-b2d5-ff9f7dacfe0f","81d11171-5d9e-4950-98f8-b0ab0f2d7908","65ddccd2-70a7-48b3-88e1-9d6fd1dcd11b","8c0d361f-eaab-4768-90f2-b7a610e2bbb1"],["TCGA-ER-A19T","TCGA-04-1655-01A-01R-1566-13","TCGA-13-0913-02A-01R-1564-13","TCGA-AB-2936","TCGA-33-4589","TCGA-91-7771","TCGA-CS-6665","TCGA-36-2539","TCGA-10-0934-01A-02R-1564-13","TCGA-2A-AAYO-10A-01D-A41M-01"],["TCGA-ER-A19T","TCGA-04-1655","TCGA-13-0913","TCGA-AB-2936","TCGA-33-4589","TCGA-91-7771","TCGA-CS-6665","TCGA-36-2539","TCGA-10-0934","TCGA-2A-AAYO"]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th> \u003c/th>\n      <th>status\u003c/th>\n      <th>category\u003c/th>\n      <th>entity_id\u003c/th>\n      <th>classification\u003c/th>\n      <th>updated_datetime\u003c/th>\n      <th>created_datetime\u003c/th>\n      <th>notes\u003c/th>\n      <th>entity_type\u003c/th>\n      <th>submitter_id\u003c/th>\n      <th>case_id\u003c/th>\n      <th>entity_submitter_id\u003c/th>\n      <th>case_submitter_id\u003c/th>\n    \u003c/tr>\n  \u003c/thead>\n\u003c/table>","options":{"order":[],"autoWidth":false,"orderClasses":false,"columnDefs":[{"orderable":false,"targets":0}]}},"evals":[],"jsHooks":[]}</script></div>
</div>
<div id="advanced-api-usage" class="section level1">
<h1 class="hasAnchor"><html><body><a href="#advanced-api-usage" class="anchor"> </a></body></html>Advanced API Usage</h1>
<div id="searching-and-filtering" class="section level2">
<h2 class="hasAnchor"><html><body><a href="#searching-and-filtering" class="anchor"> </a></body></html>Searching and filtering</h2>
</div>
<div id="faceting-counting" class="section level2">
<h2 class="hasAnchor"><html><body><a href="#faceting-counting" class="anchor"> </a></body></html>Faceting (counting)</h2>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">res =<span class="st"> </span><span class="kw"><a href="../reference/case_fields.html">cases</a></span>(<span class="dt">facets=</span><span class="st">'diagnoses.tumor_stage'</span>)</code></pre></div>
<pre><code>## GET request uri:
## https://gdc-api.nci.nih.gov/cases?format=JSON&amp;pretty=FALSE&amp;fields=sample_ids,portion_ids,submitter_portion_ids,created_datetime,submitter_aliquot_ids,updated_datetime,submitter_analyte_ids,analyte_ids,submitter_id,case_id,state,aliquot_ids,slide_ids,submitter_sample_ids&amp;size=10&amp;from=1&amp;facets=diagnoses.tumor_stage</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">a =<span class="st"> </span><span class="kw"><a href="../reference/mapping.html">mapping</a></span>(<span class="st">'projects'</span>)</code></pre></div>
<pre><code>## GET request uri:
## https://gdc-api.nci.nih.gov/projects/_mapping?format=JSON&amp;pretty=FALSE&amp;size=10&amp;from=1</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">a               <span class="co"># available information</span></code></pre></div>
<pre><code>## class: mapping_list
## names:
##     _mapping, defaults, expand, fields, multi, nested</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">a[<span class="st">"defaults"</span>]   <span class="co"># default fields</span></code></pre></div>
<pre><code>## class: mapping_list
## name: defaults
## values:
##     project_id, disease_type, released, state, primary_site,
##     dbgap_accession_number, name</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">a[[<span class="st">"defaults"</span>]] <span class="co"># default fields as character()</span></code></pre></div>
<pre><code>## [1] "project_id"             "disease_type"          
## [3] "released"               "state"                 
## [5] "primary_site"           "dbgap_accession_number"
## [7] "name"</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">a[<span class="st">"fields"</span>]     <span class="co"># all fields</span></code></pre></div>
<pre><code>## class: mapping_list
## name: fields
## values:
##     project_id, disease_type, summary.data_categories.case_count,
##     summary.data_categories.file_count,
##     summary.data_categories.data_category, summary.case_count,
##     summary.file_count,
##     summary.experimental_strategies.case_count,
##     summary.experimental_strategies.file_count,
##     summary.experimental_strategies.experimental_strategy,
##     summary.file_size, released, state,
##     program.dbgap_accession_number, program.name,
##     program.program_id, primary_site, dbgap_accession_number, name</code></pre>
</div>
</div>
<div id="advanced-querying-with-filters" class="section level1">
<h1 class="hasAnchor"><html><body><a href="#advanced-querying-with-filters" class="anchor"> </a></body></html>Advanced querying with filters</h1>
<p>The GDC supports <a href="https://docs.gdc.cancer.gov/Data_Portal/Users_Guide/Advanced_Search/">advanced queries</a> based on filters and facets. The GenomicDataCommons package allows writing queries using standard R syntax, simplifying the creation of complicated filtering patterns. Examples follow:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># youngcases = cases(filters=filters(diagnoses.age_at_diagnosis &lt;= 10*365,endpoint='cases'))</span></code></pre></div>
</div>
<div id="sessioninfo" class="section level1">
<h1 class="hasAnchor"><html><body><a href="#sessioninfo" class="anchor"> </a></body></html>SessionInfo</h1>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">sessionInfo</span>()</code></pre></div>
<pre><code>## R Under development (unstable) (2016-10-26 r71594)
## Platform: x86_64-apple-darwin13.4.0 (64-bit)
## Running under: macOS Sierra 10.12.1
## 
## locale:
## [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8
## 
## attached base packages:
## [1] stats     graphics  grDevices utils     datasets  methods   base     
## 
## other attached packages:
## [1] GenomicDataCommons_0.1.4 knitr_1.15.1            
## [3] BiocStyle_2.3.15         jsonlite_1.1            
## [5] httr_1.2.1               lazyeval_0.2.0          
## [7] DT_0.2                  
## 
## loaded via a namespace (and not attached):
##  [1] Rcpp_0.12.8.2        whisker_0.3-2        xml2_1.0.0.9001     
##  [4] magrittr_1.5         roxygen2_5.0.1       devtools_1.12.0.9000
##  [7] pkgload_0.0.0.9000   R6_2.2.0             stringr_1.1.0       
## [10] tcltk_3.4.0          tools_3.4.0          pkgbuild_0.0.0.9000 
## [13] withr_1.0.2          htmltools_0.3.5      assertthat_0.1      
## [16] yaml_2.1.14          rprojroot_1.1        digest_0.6.10       
## [19] tibble_1.2           crayon_1.3.2         pkgdown_0.1.0.9000  
## [22] purrr_0.2.2          htmlwidgets_0.8      curl_2.3            
## [25] memoise_1.0.0        evaluate_0.10        rmarkdown_1.2.9000  
## [28] stringi_1.1.2        compiler_3.4.0       desc_1.0.1          
## [31] backports_1.0.4      highlight_0.4.7</code></pre>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2>Contents</h2>
      <ul class="nav nav-pills nav-stacked"><li><a href="#background">Background</a><ul class="nav nav-pills nav-stacked"><li><a href="#data-model">Data Model</a></li>
      <li><a href="#full-feature-set-of-the-gdc-api">Full Feature Set of the GDC API</a></li>
      </ul></li>
      <li><a href="#the-basic-api-endpoints">The basic API endpoints</a><ul class="nav nav-pills nav-stacked"><li><a href="#projects">Projects</a></li>
      <li><a href="#files">Files</a></li>
      <li><a href="#cases">Cases</a></li>
      <li><a href="#annotations">Annotations</a></li>
      </ul></li>
      <li><a href="#advanced-api-usage">Advanced API Usage</a><ul class="nav nav-pills nav-stacked"><li><a href="#searching-and-filtering">Searching and filtering</a></li>
      <li><a href="#faceting-counting">Faceting (counting)</a></li>
      </ul></li>
      <li><a href="#advanced-querying-with-filters">Advanced querying with filters</a></li>
      <li><a href="#sessioninfo">SessionInfo</a></li>
      </ul></div>
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by Martin Morgan, Davis Sean.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://hadley.github.io/pkgdown/">pkgdown</a>.</p>
</div>

      </footer></div>

  </body></html>
