<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html>
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Query GDC for data slices — slicing • GenomicDataCommons</title>

<!-- jquery -->
<script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script>
<!-- Bootstrap -->
<link href="https://maxcdn.bootstrapcdn.com/bootswatch/3.3.7/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">

<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">


<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../jquery.sticky-kit.min.js"></script>
<script src="../pkgdown.js"></script>

<!-- mathjax -->
<script src='https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML'></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


<!-- Google analytics -->
<script>
  (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
  (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
  m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
  })(window,document,'script','https://www.google-analytics.com/analytics.js','ga');

  ga('create', 'UA-93043521-1', 'auto');
  ga('send', 'pageview');

</script>

  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="../index.html">GenomicDataCommons</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../articles/index.html">Articles</a>
</li>
<li>
  <a href="../news/index.html">News</a>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="http://github.com/Bioconductor/GenomicDataCommons">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

      <div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Query GDC for data slices</h1>
    </div>

    
    <p>This function returns a BAM file representing reads overlapping
regions specified either as chromosomal regions or as gencode gene
symbols.</p>
    

    <pre><span class='fu'>slicing</span>(<span class='no'>uuid</span>, <span class='no'>regions</span>, <span class='no'>symbols</span>, <span class='kw'>destination</span> <span class='kw'>=</span> <span class='fu'>tempfile</span>(), <span class='kw'>overwrite</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>,
  <span class='kw'>progress</span> <span class='kw'>=</span> <span class='fu'>interactive</span>(), <span class='kw'>token</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>legacy</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>)</pre>
    
    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a> Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>uuid</th>
      <td><p>character(1) identifying the BAM file resource</p></td>
    </tr>
    <tr>
      <th>regions</th>
      <td><p>character() vector describing chromosomal regions,
e.g., <code>c("chr1", "chr2:10000", "chr3:10000-20000")</code> (all
of chromosome 1, chromosome 2 from position 10000 to the end,
chromosome 3 from 10000 to 20000).</p></td>
    </tr>
    <tr>
      <th>symbols</th>
      <td><p>character() vector of gencode gene symbols, e.g.,
<code>c("BRCA1", "PTEN")</code></p></td>
    </tr>
    <tr>
      <th>destination</th>
      <td><p>character(1) default <code>tempfile()</code> file path
for BAM file slice</p></td>
    </tr>
    <tr>
      <th>overwrite</th>
      <td><p>logical(1) default FALSE can destination be
overwritten?</p></td>
    </tr>
    <tr>
      <th>progress</th>
      <td><p>logical(1) default <code>interactive()</code> should a
progress bar be used?</p></td>
    </tr>
    <tr>
      <th>token</th>
      <td><p>character(1) security token allowing access to
restricted data. Almost all BAM data is restricted, so a token is
usually required. See
<a href = 'https://gdc-docs.nci.nih.gov/API/Users_Guide/Authentication_and_Authorization/'>https://gdc-docs.nci.nih.gov/API/Users_Guide/Authentication_and_Authorization/</a>.</p></td>
    </tr>
    <tr>
      <th>legacy</th>
      <td><p>logical(1) whether or not to use the "legacy"
archive, containing older, non-harmonized data.</p></td>
    </tr>
    </table>
    
    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>character(1) destination to the downloaded BAM file</p>
    

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'>
<span class='fu'>slicing</span>(<span class='st'>"df80679e-c4d3-487b-934c-fcc782e5d46e"</span>,
        <span class='kw'>regions</span><span class='kw'>=</span><span class='st'>"chr17:75000000-76000000"</span>,
        <span class='kw'>token</span><span class='kw'>=</span><span class='fu'><a href='gdc_token.html'>gdc_token</a></span>())</div><div class='output co'>#&gt; <span class='error'>Error in .gdc_post(endpoint = sprintf("legacy/slicing/view/%s", uuid),     add_headers(`Content-type` = "application/json"), write_disk(destination,         overwrite), if (progress) progress() else NULL, body = toJSON(body),     token = token, legacy = legacy): Not Found (HTTP 404).</span></div><div class='input'>
</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      
      <li><a href="#value">Value</a></li>
      
      <li><a href="#examples">Examples</a></li>
    </ul>

  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by Martin Morgan, Davis Sean.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://hadley.github.io/pkgdown/">pkgdown</a>.</p>
</div>

      </footer>
   </div>

  </body>
</html>
