<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>The GenomicDataCommons Package • GenomicDataCommons</title>
<!-- jquery --><script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://maxcdn.bootstrapcdn.com/bootswatch/3.3.7/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script><!-- Font Awesome icons --><link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/1.7.1/clipboard.min.js" integrity="sha384-cV+rhyOuRHc9Ub/91rihWcGmMmCXDeksTtCihMupQHSsi8GIIRDG0ThDc3HGQFJ3" crossorigin="anonymous"></script><!-- sticky kit --><script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="The GenomicDataCommons Package">
<meta property="og:description" content="">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=UA-93043521-1"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-93043521-1');
</script>
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">GenomicDataCommons</a>
        <span class="label label-default" data-toggle="tooltip" data-placement="bottom" title="Released package">1.5.5</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/overview.html">The GenomicDataCommons Package</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/Bioconductor/GenomicDataCommons">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>The GenomicDataCommons Package</h1>
                        <h4 class="author">Sean Davis &amp; Martin Morgan</h4>
            
            <h4 class="date">Tuesday, September 04, 2018</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/Bioconductor/GenomicDataCommons/blob/master/vignettes/overview.Rmd"><code>vignettes/overview.Rmd</code></a></small>
      <div class="hidden name"><code>overview.Rmd</code></div>

    </div>

    
        <div class="abstract">
      <p class="abstract">Abstract</p>
      The National Cancer Institute (NCI) has established the <a href="https://gdc.nci.nih.gov/">Genomic Data Commons</a> (GDC). The GDC provides the cancer research community with an open and unified repository for sharing and accessing data across numerous cancer studies and projects via a high-performance data transfer and query infrastructure. The <em>GenomicDataCommons</em> Bioconductor package provides basic infrastructure for querying, accessing, and mining genomic datasets available from the GDC. We expect that the Bioconductor developer and the larger bioinformatics communities will build on the <em>GenomicDataCommons</em> package to add higher-level functionality and expose cancer genomics data to the plethora of state-of-the-art bioinformatics methods available in Bioconductor.
    </div>
    
<div id="what-is-the-gdc" class="section level1">
<h1 class="hasAnchor">
<a href="#what-is-the-gdc" class="anchor"></a>What is the GDC?</h1>
<p>From the <a href="https://gdc.cancer.gov/about-gdc">Genomic Data Commons (GDC) website</a>:</p>
<blockquote>
<p>The National Cancer Institute’s (NCI’s) Genomic Data Commons (GDC) is a data sharing platform that promotes precision medicine in oncology. It is not just a database or a tool; it is an expandable knowledge network supporting the import and standardization of genomic and clinical data from cancer research programs. The GDC contains NCI-generated data from some of the largest and most comprehensive cancer genomic datasets, including The Cancer Genome Atlas (TCGA) and Therapeutically Applicable Research to Generate Effective Therapies (TARGET). For the first time, these datasets have been harmonized using a common set of bioinformatics pipelines, so that the data can be directly compared. As a growing knowledge system for cancer, the GDC also enables researchers to submit data, and harmonizes these data for import into the GDC. As more researchers add clinical and genomic data to the GDC, it will become an even more powerful tool for making discoveries about the molecular basis of cancer that may lead to better care for patients.</p>
</blockquote>
<p>The <a href="https://gdc.cancer.gov/developers/gdc-data-model/gdc-data-model-components">data model for the GDC is complex</a>, but it worth a quick overview and a graphical representation is included here.</p>
<div class="figure">
<img src="all_nodes_040318.png" alt="The data model is encoded as a so-called property graph. Nodes represent entities such as Projects, Cases, Diagnoses, Files (various kinds), and Annotations. The relationships between these entities are maintained as edges. Both nodes and edges may have Properties that supply instance details."><p class="caption">The data model is encoded as a so-called property graph. Nodes represent entities such as Projects, Cases, Diagnoses, Files (various kinds), and Annotations. The relationships between these entities are maintained as edges. Both nodes and edges may have Properties that supply instance details.</p>
</div>
<p>The GDC API exposes these nodes and edges in a somewhat simplified set of <a href="https://en.wikipedia.org/wiki/Representational_state_transfer">RESTful</a> endpoints.</p>
</div>
<div id="quickstart" class="section level1">
<h1 class="hasAnchor">
<a href="#quickstart" class="anchor"></a>Quickstart</h1>
<p>This quickstart section is just meant to show basic functionality. More details of functionality are included further on in this vignette and in function-specific help.</p>
<p>This software is available at Bioconductor.org and can be downloaded via <code><a href="http://www.rdocumentation.org/packages/BiocManager/topics/install">BiocManager::install</a></code>.</p>
<p>To report bugs or problems, either <a href="https://github.com/Bioconductor/GenomicDataCommons/issues">submit a new issue</a> or submit a <code>bug.report(package='GenomicDataCommons')</code> from within R (which will redirect you to the new issue on GitHub).</p>
<div id="installation" class="section level2">
<h2 class="hasAnchor">
<a href="#installation" class="anchor"></a>Installation</h2>
<p>Installation can be achieved via Bioconductor’s <code>BiocManager</code> package.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb1-1" data-line-number="1"><span class="cf">if</span> (<span class="op">!</span><span class="kw">require</span>(<span class="st">"BiocManager"</span>))</a>
<a class="sourceLine" id="cb1-2" data-line-number="2">    <span class="kw">install.packages</span>(<span class="st">"BiocManager"</span>)</a>
<a class="sourceLine" id="cb1-3" data-line-number="3">BiocManager<span class="op">::</span><span class="kw"><a href="http://www.rdocumentation.org/packages/BiocManager/topics/install">install</a></span>(<span class="st">'GenomicDataCommons'</span>)</a></code></pre></div>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb2-1" data-line-number="1"><span class="kw">library</span>(GenomicDataCommons)</a></code></pre></div>
</div>
<div id="check-connectivity-and-status" class="section level2">
<h2 class="hasAnchor">
<a href="#check-connectivity-and-status" class="anchor"></a>Check connectivity and status</h2>
<p>The <em><a href="http://bioconductor.org/packages/GenomicDataCommons">GenomicDataCommons</a></em> package relies on having network connectivity. In addition, the NCI GDC API must also be operational and not under maintenance. Checking <code>status</code> can be used to check this connectivity and functionality.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb3-1" data-line-number="1">GenomicDataCommons<span class="op">::</span><span class="kw"><a href="http://www.rdocumentation.org/packages/GenomicDataCommons/topics/status">status</a></span>()</a></code></pre></div>
<pre><code>## $commit
## [1] "8acb9697bd3e944def954f8d1c0b5c22a0fbc8fa"
## 
## $data_release
## [1] "Data Release 12.0 - August 23, 2018"
## 
## $status
## [1] "OK"
## 
## $tag
## [1] "1.15.1"
## 
## $version
## [1] 1</code></pre>
<p>And to check the status in code:</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb5-1" data-line-number="1"><span class="kw">stopifnot</span>(GenomicDataCommons<span class="op">::</span><span class="kw"><a href="http://www.rdocumentation.org/packages/GenomicDataCommons/topics/status">status</a></span>()<span class="op">$</span>status<span class="op">==</span><span class="st">"OK"</span>)</a></code></pre></div>
</div>
<div id="find-data" class="section level2">
<h2 class="hasAnchor">
<a href="#find-data" class="anchor"></a>Find data</h2>
<p>The following code builds a <code>manifest</code> that can be used to guide the download of raw data. Here, filtering finds gene expression files quantified as raw counts using <code>HTSeq</code> from ovarian cancer patients.</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb6-1" data-line-number="1">ge_manifest =<span class="st"> </span><span class="kw"><a href="../reference/query.html">files</a></span>() <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb6-2" data-line-number="2"><span class="st">    </span><span class="kw"><a href="../reference/filtering.html">filter</a></span>( <span class="op">~</span><span class="st"> </span>cases.project.project_id <span class="op">==</span><span class="st"> 'TCGA-OV'</span> <span class="op">&amp;</span></a>
<a class="sourceLine" id="cb6-3" data-line-number="3"><span class="st">                </span>type <span class="op">==</span><span class="st"> 'gene_expression'</span> <span class="op">&amp;</span></a>
<a class="sourceLine" id="cb6-4" data-line-number="4"><span class="st">                </span>analysis.workflow_type <span class="op">==</span><span class="st"> 'HTSeq - Counts'</span>) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb6-5" data-line-number="5"><span class="st">    </span><span class="kw"><a href="../reference/manifest.html">manifest</a></span>()</a></code></pre></div>
</div>
<div id="download-data" class="section level2">
<h2 class="hasAnchor">
<a href="#download-data" class="anchor"></a>Download data</h2>
<p>After the 379 gene expression files specified in the query above. Using multiple processes to do the download very significantly speeds up the transfer in many cases. On a standard 1Gb connection, the following completes in about 30 seconds. The first time the data are downloaded, R will ask to create a cache directory (see <code><a href="../reference/gdc_cache.html">?gdc_cache</a></code> for details of setting and interacting with the cache). Resulting downloaded files will be stored in the cache directory. Future access to the same files will be directly from the cache, alleviating multiple downloads.</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb7-1" data-line-number="1">fnames =<span class="st"> </span><span class="kw">lapply</span>(ge_manifest<span class="op">$</span>id[<span class="dv">1</span><span class="op">:</span><span class="dv">20</span>],gdcdata)</a></code></pre></div>
<p>If the download had included controlled-access data, the download above would have needed to include a <code>token</code>. Details are available in <a href="#authentication">the authentication section below</a>.</p>
</div>
<div id="metadata-queries" class="section level2">
<h2 class="hasAnchor">
<a href="#metadata-queries" class="anchor"></a>Metadata queries</h2>
<p>The <em><a href="http://bioconductor.org/packages/GenomicDataCommons">GenomicDataCommons</a></em> can access the significant clinical, demographic, biospecimen, and annotation information contained in the NCI GDC.</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb8-1" data-line-number="1">expands =<span class="st"> </span><span class="kw">c</span>(<span class="st">"diagnoses"</span>,<span class="st">"annotations"</span>,</a>
<a class="sourceLine" id="cb8-2" data-line-number="2">             <span class="st">"demographic"</span>,<span class="st">"exposures"</span>)</a>
<a class="sourceLine" id="cb8-3" data-line-number="3">clinResults =<span class="st"> </span><span class="kw"><a href="../reference/query.html">cases</a></span>() <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb8-4" data-line-number="4"><span class="st">    </span>GenomicDataCommons<span class="op">::</span><span class="kw"><a href="http://www.rdocumentation.org/packages/GenomicDataCommons/topics/select">select</a></span>(<span class="ot">NULL</span>) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb8-5" data-line-number="5"><span class="st">    </span>GenomicDataCommons<span class="op">::</span><span class="kw"><a href="http://www.rdocumentation.org/packages/GenomicDataCommons/topics/expand">expand</a></span>(expands) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb8-6" data-line-number="6"><span class="st">    </span><span class="kw"><a href="../reference/results.html">results</a></span>(<span class="dt">size=</span><span class="dv">50</span>)</a>
<a class="sourceLine" id="cb8-7" data-line-number="7"><span class="kw">str</span>(clinResults[[<span class="dv">1</span>]],<span class="dt">list.len=</span><span class="dv">10</span>)</a></code></pre></div>
<pre><code>## List of 50
##  $ 1f95a9f1-0327-4c89-a6fe-69809d0d8ebf:'data.frame':    1 obs. of  19 variables:
##   ..$ progression_or_recurrence        : chr "not reported"
##   ..$ classification_of_tumor          : chr "primary"
##   ..$ last_known_disease_status        : chr "not reported"
##   ..$ tumor_grade                      : chr "not reported"
##   ..$ tissue_or_organ_of_origin        : chr "Breast, NOS"
##   ..$ created_datetime                 : chr "2017-06-16T15:41:54.034178-05:00"
##   ..$ updated_datetime                 : chr "2018-08-07T05:08:25.415044+00:00"
##   ..$ primary_diagnosis                : chr "Infiltrating duct carcinoma, NOS"
##   ..$ submitter_id                     : chr "AD11218_diagnosis"
##   ..$ site_of_resection_or_biopsy      : chr "Breast, NOS"
##   .. [list output truncated]
##  $ 4060f92b-ab42-4233-8609-6e9baa11631d:'data.frame':    1 obs. of  19 variables:
##   ..$ progression_or_recurrence        : chr "not reported"
##   ..$ classification_of_tumor          : chr "metastasis"
##   ..$ last_known_disease_status        : chr "not reported"
##   ..$ tumor_grade                      : chr "not reported"
##   ..$ tissue_or_organ_of_origin        : chr "Nervous system, NOS"
##   ..$ created_datetime                 : chr "2017-06-19T09:23:29.064224-05:00"
##   ..$ updated_datetime                 : chr "2018-08-07T05:08:25.415044+00:00"
##   ..$ primary_diagnosis                : chr "Oligodendroglioma, NOS"
##   ..$ submitter_id                     : chr "AD8623_diagnosis"
##   ..$ site_of_resection_or_biopsy      : chr "Brain, NOS"
##   .. [list output truncated]
##  $ 90f4b65c-cfd4-4066-a5b9-842885c172e2:'data.frame':    1 obs. of  22 variables:
##   ..$ classification_of_tumor          : chr "not reported"
##   ..$ last_known_disease_status        : chr "not reported"
##   ..$ updated_datetime                 : chr "2018-03-09T16:42:15.753583-06:00"
##   ..$ primary_diagnosis                : chr "C56.9"
##   ..$ submitter_id                     : chr "TCGA-13-0727_diagnosis"
##   ..$ tumor_stage                      : chr "not reported"
##   ..$ age_at_diagnosis                 : int 26185
##   ..$ vital_status                     : chr "dead"
##   ..$ morphology                       : chr "8441/3"
##   ..$ days_to_death                    : num 462
##   .. [list output truncated]
##  $ a4ec8b0b-0b53-43e6-90b4-e88a8a3cbc10:'data.frame':    1 obs. of  19 variables:
##   ..$ progression_or_recurrence        : chr "not reported"
##   ..$ classification_of_tumor          : chr "primary"
##   ..$ last_known_disease_status        : chr "not reported"
##   ..$ tumor_grade                      : chr "not reported"
##   ..$ tissue_or_organ_of_origin        : chr "Breast, NOS"
##   ..$ created_datetime                 : chr "2017-06-19T09:29:52.147488-05:00"
##   ..$ updated_datetime                 : chr "2018-08-07T05:08:25.415044+00:00"
##   ..$ primary_diagnosis                : chr "Infiltrating duct carcinoma, NOS"
##   ..$ submitter_id                     : chr "AD1214_diagnosis"
##   ..$ site_of_resection_or_biopsy      : chr "Breast, NOS"
##   .. [list output truncated]
##  $ 6e534dd3-81ac-4575-ab1e-02c21e10916d:'data.frame':    1 obs. of  22 variables:
##   ..$ classification_of_tumor          : chr "not reported"
##   ..$ last_known_disease_status        : chr "not reported"
##   ..$ updated_datetime                 : chr "2018-03-08T19:30:04.867636-06:00"
##   ..$ primary_diagnosis                : chr "C24.0"
##   ..$ submitter_id                     : chr "TCGA-W5-AA2X_diagnosis"
##   ..$ tumor_stage                      : chr "stage ivb"
##   ..$ age_at_diagnosis                 : int 24797
##   ..$ vital_status                     : chr "dead"
##   ..$ morphology                       : chr "8160/3"
##   ..$ days_to_death                    : num 271
##   .. [list output truncated]
##  $ dfd8aaad-cf31-5260-8882-b471ddbcc8fd:'data.frame':    1 obs. of  17 variables:
##   ..$ progression_or_recurrence  : chr "not reported"
##   ..$ last_known_disease_status  : chr "not reported"
##   ..$ tumor_grade                : chr "Not Reported"
##   ..$ tissue_or_organ_of_origin  : chr "Not Reported"
##   ..$ created_datetime           : chr "2017-02-24T20:54:45.808091-06:00"
##   ..$ updated_datetime           : chr "2018-01-19T13:39:21.801433-06:00"
##   ..$ primary_diagnosis          : chr "not reported"
##   ..$ submitter_id               : chr "TARGET-50-PADXJZ_diagnosis"
##   ..$ site_of_resection_or_biopsy: chr "C649"
##   ..$ tumor_stage                : chr "iii"
##   .. [list output truncated]
##  $ 72636f64-ba2c-512e-b76f-72ff91f7f0f7:'data.frame':    1 obs. of  17 variables:
##   ..$ progression_or_recurrence  : chr "not reported"
##   ..$ last_known_disease_status  : chr "not reported"
##   ..$ tumor_grade                : chr "Not Reported"
##   ..$ tissue_or_organ_of_origin  : chr "Not Reported"
##   ..$ created_datetime           : chr "2017-02-24T21:09:16.860616-06:00"
##   ..$ updated_datetime           : chr "2018-03-08T09:11:04.174187-06:00"
##   ..$ primary_diagnosis          : chr "not reported"
##   ..$ submitter_id               : chr "TARGET-20-PARPEK_diagnosis"
##   ..$ site_of_resection_or_biopsy: chr "not reported"
##   ..$ tumor_stage                : chr "not reported"
##   .. [list output truncated]
##  $ 2fb5e38a-6b65-43ae-945c-9fa8109b2de5:'data.frame':    1 obs. of  19 variables:
##   ..$ progression_or_recurrence        : chr "not reported"
##   ..$ classification_of_tumor          : chr "primary"
##   ..$ last_known_disease_status        : chr "not reported"
##   ..$ tumor_grade                      : chr "not reported"
##   ..$ tissue_or_organ_of_origin        : chr "Rectum, NOS"
##   ..$ created_datetime                 : chr "2017-06-19T09:11:37.838594-05:00"
##   ..$ updated_datetime                 : chr "2018-08-07T05:08:25.415044+00:00"
##   ..$ primary_diagnosis                : chr "Squamous cell carcinoma, NOS"
##   ..$ submitter_id                     : chr "AD6701_diagnosis"
##   ..$ site_of_resection_or_biopsy      : chr "Rectum, NOS"
##   .. [list output truncated]
##  $ c34c0e02-2285-448f-bf8a-433e4caf5379:'data.frame':    1 obs. of  19 variables:
##   ..$ progression_or_recurrence        : chr "not reported"
##   ..$ classification_of_tumor          : chr "primary"
##   ..$ last_known_disease_status        : chr "not reported"
##   ..$ tumor_grade                      : chr "not reported"
##   ..$ tissue_or_organ_of_origin        : chr "Lung, NOS"
##   ..$ created_datetime                 : chr "2017-06-16T16:06:02.946257-05:00"
##   ..$ updated_datetime                 : chr "2018-08-07T05:08:25.415044+00:00"
##   ..$ primary_diagnosis                : chr "Adenocarcinoma, NOS"
##   ..$ submitter_id                     : chr "AD4438_diagnosis"
##   ..$ site_of_resection_or_biopsy      : chr "Lung, NOS"
##   .. [list output truncated]
##  $ ee3fa9cd-2d82-4bc1-b71a-87818045dd70:'data.frame':    1 obs. of  19 variables:
##   ..$ progression_or_recurrence        : chr "not reported"
##   ..$ classification_of_tumor          : chr "primary"
##   ..$ last_known_disease_status        : chr "not reported"
##   ..$ tumor_grade                      : chr "not reported"
##   ..$ tissue_or_organ_of_origin        : chr "Lung, NOS"
##   ..$ created_datetime                 : chr "2017-06-16T15:39:04.464841-05:00"
##   ..$ updated_datetime                 : chr "2018-08-07T05:08:25.415044+00:00"
##   ..$ primary_diagnosis                : chr "Squamous cell carcinoma, NOS"
##   ..$ submitter_id                     : chr "AD9891_diagnosis"
##   ..$ site_of_resection_or_biopsy      : chr "Lung, NOS"
##   .. [list output truncated]
##   [list output truncated]</code></pre>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb10-1" data-line-number="1"><span class="co"># or listviewer::jsonedit(clinResults)</span></a></code></pre></div>
</div>
</div>
<div id="basic-design" class="section level1">
<h1 class="hasAnchor">
<a href="#basic-design" class="anchor"></a>Basic design</h1>
<p>This package design is meant to have some similarities to the “hadleyverse” approach of dplyr. Roughly, the functionality for finding and accessing files and metadata can be divided into:</p>
<ol style="list-style-type: decimal">
<li>Simple query constructors based on GDC API endpoints.</li>
<li>A set of verbs that when applied, adjust filtering, field selection, and faceting (fields for aggregation) and result in a new query object (an endomorphism)</li>
<li>A set of verbs that take a query and return results from the GDC</li>
</ol>
<p>In addition, there are exhiliary functions for asking the GDC API for information about available and default fields, slicing BAM files, and downloading actual data files. Here is an overview of functionality<a href="#fn1" class="footnote-ref" id="fnref1"><sup>1</sup></a>.</p>
<ul>
<li>Creating a query
<ul>
<li><code><a href="../reference/query.html">projects()</a></code></li>
<li><code><a href="../reference/query.html">cases()</a></code></li>
<li><code><a href="../reference/query.html">files()</a></code></li>
<li><code><a href="../reference/query.html">annotations()</a></code></li>
</ul>
</li>
<li>Manipulating a query
<ul>
<li><code><a href="../reference/filtering.html">filter()</a></code></li>
<li><code><a href="../reference/faceting.html">facet()</a></code></li>
<li><code><a href="../reference/select.html">select()</a></code></li>
</ul>
</li>
<li>Introspection on the GDC API fields
<ul>
<li><code><a href="../reference/mapping.html">mapping()</a></code></li>
<li><code><a href="../reference/available_fields.html">available_fields()</a></code></li>
<li><code><a href="../reference/default_fields.html">default_fields()</a></code></li>
<li><code><a href="../reference/grep_fields.html">grep_fields()</a></code></li>
<li><code>field_picker()</code></li>
<li><code><a href="../reference/available_values.html">available_values()</a></code></li>
<li><code><a href="../reference/available_expand.html">available_expand()</a></code></li>
</ul>
</li>
<li>Executing an API call to retrieve query results
<ul>
<li><code><a href="../reference/results.html">results()</a></code></li>
<li><code><a href="../reference/count.html">count()</a></code></li>
<li><code><a href="../reference/response.html">response()</a></code></li>
</ul>
</li>
<li>Raw data file downloads
<ul>
<li><code><a href="../reference/gdcdata.html">gdcdata()</a></code></li>
<li><code><a href="../reference/transfer.html">transfer()</a></code></li>
<li><code><a href="../reference/gdc_client.html">gdc_client()</a></code></li>
</ul>
</li>
<li>Summarizing and aggregating field values (faceting)
<ul>
<li><code><a href="../reference/aggregations.html">aggregations()</a></code></li>
</ul>
</li>
<li>Authentication
<ul>
<li><code><a href="../reference/gdc_token.html">gdc_token()</a></code></li>
</ul>
</li>
<li>BAM file slicing
<ul>
<li><code><a href="../reference/slicing.html">slicing()</a></code></li>
</ul>
</li>
</ul>
</div>
<div id="usage" class="section level1">
<h1 class="hasAnchor">
<a href="#usage" class="anchor"></a>Usage</h1>
<p>There are two main classes of operations when working with the NCI GDC.</p>
<ol style="list-style-type: decimal">
<li>
<a href="#querying-metadata">Querying metadata and finding data files</a> (e.g., finding all gene expression quantifications data files for all colon cancer patients).</li>
<li>
<a href="#datafile-access-and-download">Transferring raw or processed data</a> from the GDC to another computer (e.g., downloading raw or processed data)</li>
</ol>
<p>Both classes of operation are reviewed in detail in the following sections.</p>
<div id="querying-metadata" class="section level2">
<h2 class="hasAnchor">
<a href="#querying-metadata" class="anchor"></a>Querying metadata</h2>
<p>Vast amounts of metadata about cases (patients, basically), files, projects, and so-called annotations are available via the NCI GDC API. Typically, one will want to query metadata to either focus in on a set of files for download or transfer <em>or</em> to perform so-called aggregations (pivot-tables, facets, similar to the R <code>table()</code> functionality).</p>
<p>Querying metadata starts with <a href="#creating-a-query">creating a “blank” query</a>. One will often then want to <a href="#filtering"><code>filter</code></a> the query to limit results prior to <a href="#retrieving-results">retrieving results</a>. The GenomicDataCommons package has <a href="#fields-and-values">helper functions for listing fields</a> that are available for filtering.</p>
<p>In addition to fetching results, the GDC API allows <a href="#facets-and-aggregation">faceting, or aggregating,</a>, useful for compiling reports, generating dashboards, or building user interfaces to GDC data (see GDC web query interface for a non-R-based example).</p>
<div id="creating-a-query" class="section level3">
<h3 class="hasAnchor">
<a href="#creating-a-query" class="anchor"></a>Creating a query</h3>
<p>A query of the GDC starts its life in R. Queries follow the four metadata endpoints available at the GDC. In particular, there are four convenience functions that each create <code>GDCQuery</code> objects (actually, specific subclasses of <code>GDCQuery</code>):</p>
<ul>
<li><code><a href="../reference/query.html">projects()</a></code></li>
<li><code><a href="../reference/query.html">cases()</a></code></li>
<li><code><a href="../reference/query.html">files()</a></code></li>
<li><code><a href="../reference/query.html">annotations()</a></code></li>
</ul>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb11-1" data-line-number="1">pquery =<span class="st"> </span><span class="kw"><a href="../reference/query.html">projects</a></span>()</a></code></pre></div>
<p>The <code>pquery</code> object is now an object of (S3) class, <code>GDCQuery</code> (and <code>gdc_projects</code> and <code>list</code>). The object contains the following elements:</p>
<ul>
<li>fields: This is a character vector of the fields that will be returned when we <a href="#retrieving-results">retrieve data</a>. If no fields are specified to, for example, the <code><a href="../reference/query.html">projects()</a></code> function, the default fields from the GDC are used (see <code><a href="../reference/default_fields.html">default_fields()</a></code>)</li>
<li>filters: This will contain results after calling the <a href="#filtering"><code><a href="../reference/filtering.html">filter()</a></code> method</a> and will be used to filter results on <a href="#retrieving-results">retrieval</a>.</li>
<li>facets: A character vector of field names that will be used for <a href="#facets-and-aggregation">aggregating data</a> in a call to <code><a href="../reference/aggregations.html">aggregations()</a></code>.</li>
<li>archive: One of either “default” or <a href="https://gdc-portal.nci.nih.gov/legacy-archive/">“legacy”</a>.</li>
<li>token: A character(1) token from the GDC. See <a href="#authentication">the authentication section</a> for details, but note that, in general, the token is not necessary for metadata query and retrieval, only for actual data download.</li>
</ul>
<p>Looking at the actual object (get used to using <code>str()</code>!), note that the query contains no results.</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb12-1" data-line-number="1"><span class="kw">str</span>(pquery)</a></code></pre></div>
<pre><code>## List of 5
##  $ fields : chr [1:10] "dbgap_accession_number" "disease_type" "intended_release_date" "name" ...
##  $ filters: NULL
##  $ facets : NULL
##  $ legacy : logi FALSE
##  $ expand : NULL
##  - attr(*, "class")= chr [1:3] "gdc_projects" "GDCQuery" "list"</code></pre>
</div>
<div id="retrieving-results" class="section level3">
<h3 class="hasAnchor">
<a href="#retrieving-results" class="anchor"></a>Retrieving results</h3>
<p><a href="https://docs.gdc.cancer.gov/API/Users_Guide/Search_and_Retrieval/#size-and-from">[ GDC pagination documentation ]</a></p>
<p><a href="https://docs.gdc.cancer.gov/API/Users_Guide/Search_and_Retrieval/#sort">[ GDC sorting documentation ]</a></p>
<p>With a query object available, the next step is to retrieve results from the GDC. The GenomicDataCommons package. The most basic type of results we can get is a simple <code><a href="../reference/count.html">count()</a></code> of records available that satisfy the filter criteria. Note that we have not set any filters, so a <code><a href="../reference/count.html">count()</a></code> here will represent all the project records publicly available at the GDC in the “default” archive"</p>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb14-1" data-line-number="1">pcount =<span class="st"> </span><span class="kw"><a href="../reference/count.html">count</a></span>(pquery)</a>
<a class="sourceLine" id="cb14-2" data-line-number="2"><span class="co"># or</span></a>
<a class="sourceLine" id="cb14-3" data-line-number="3">pcount =<span class="st"> </span>pquery <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="../reference/count.html">count</a></span>()</a>
<a class="sourceLine" id="cb14-4" data-line-number="4">pcount</a></code></pre></div>
<pre><code>## [1] 40</code></pre>
<p>The <code><a href="../reference/results.html">results()</a></code> method will fetch actual results.</p>
<div class="sourceCode" id="cb16"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb16-1" data-line-number="1">presults =<span class="st"> </span>pquery <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="../reference/results.html">results</a></span>()</a></code></pre></div>
<p>These results are returned from the GDC in <a href="http://www.json.org/">JSON</a> format and converted into a (potentially nested) list in R. The <code>str()</code> method is useful for taking a quick glimpse of the data.</p>
<div class="sourceCode" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb17-1" data-line-number="1"><span class="kw">str</span>(presults)</a></code></pre></div>
<pre><code>## List of 9
##  $ dbgap_accession_number: chr [1:10] "phs000467" NA "phs000466" "phs000465" ...
##  $ disease_type          :List of 10
##   ..$ TARGET-NBL : chr "Neuroblastoma"
##   ..$ TCGA-SARC  : chr "Sarcoma"
##   ..$ TARGET-CCSK: chr "Clear Cell Sarcoma of the Kidney"
##   ..$ TARGET-AML : chr "Acute Myeloid Leukemia"
##   ..$ TCGA-ACC   : chr "Adrenocortical Carcinoma"
##   ..$ TCGA-LGG   : chr "Brain Lower Grade Glioma"
##   ..$ TCGA-THCA  : chr "Thyroid Carcinoma"
##   ..$ TCGA-MESO  : chr "Mesothelioma"
##   ..$ TCGA-READ  : chr "Rectum Adenocarcinoma"
##   ..$ TCGA-KIRC  : chr "Kidney Renal Clear Cell Carcinoma"
##  $ releasable            : logi [1:10] FALSE FALSE FALSE FALSE FALSE FALSE ...
##  $ released              : logi [1:10] TRUE TRUE TRUE TRUE TRUE TRUE ...
##  $ state                 : chr [1:10] "open" "open" "open" "open" ...
##  $ primary_site          :List of 10
##   ..$ TARGET-NBL : chr "Nervous System"
##   ..$ TCGA-SARC  : chr "Soft Tissue"
##   ..$ TARGET-CCSK: chr "Kidney"
##   ..$ TARGET-AML : chr "Blood"
##   ..$ TCGA-ACC   : chr "Adrenal Gland"
##   ..$ TCGA-LGG   : chr "Brain"
##   ..$ TCGA-THCA  : chr "Thyroid"
##   ..$ TCGA-MESO  : chr "Pleura"
##   ..$ TCGA-READ  : chr "Colorectal"
##   ..$ TCGA-KIRC  : chr "Kidney"
##  $ project_id            : chr [1:10] "TARGET-NBL" "TCGA-SARC" "TARGET-CCSK" "TARGET-AML" ...
##  $ id                    : chr [1:10] "TARGET-NBL" "TCGA-SARC" "TARGET-CCSK" "TARGET-AML" ...
##  $ name                  : chr [1:10] "Neuroblastoma" "Sarcoma" "Clear Cell Sarcoma of the Kidney" "Acute Myeloid Leukemia" ...
##  - attr(*, "row.names")= int [1:10] 1 2 3 4 5 6 7 8 9 10
##  - attr(*, "class")= chr [1:3] "GDCprojectsResults" "GDCResults" "list"</code></pre>
<p>A default of only 10 records are returned. We can use the <code>size</code> and <code>from</code> arguments to <code><a href="../reference/results.html">results()</a></code> to either page through results or to change the number of results. Finally, there is a convenience method, <code><a href="../reference/results_all.html">results_all()</a></code> that will simply fetch all the available results given a query. Note that <code><a href="../reference/results_all.html">results_all()</a></code> may take a long time and return HUGE result sets if not used carefully. Use of a combination of <code><a href="../reference/count.html">count()</a></code> and <code><a href="../reference/results.html">results()</a></code> to get a sense of the expected data size is probably warranted before calling <code><a href="../reference/results_all.html">results_all()</a></code></p>
<div class="sourceCode" id="cb19"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb19-1" data-line-number="1"><span class="kw">length</span>(<span class="kw"><a href="../reference/ids.html">ids</a></span>(presults))</a></code></pre></div>
<pre><code>## [1] 10</code></pre>
<div class="sourceCode" id="cb21"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb21-1" data-line-number="1">presults =<span class="st"> </span>pquery <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="../reference/results_all.html">results_all</a></span>()</a>
<a class="sourceLine" id="cb21-2" data-line-number="2"><span class="kw">length</span>(<span class="kw"><a href="../reference/ids.html">ids</a></span>(presults))</a></code></pre></div>
<pre><code>## [1] 40</code></pre>
<div class="sourceCode" id="cb23"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb23-1" data-line-number="1"><span class="co"># includes all records</span></a>
<a class="sourceLine" id="cb23-2" data-line-number="2"><span class="kw">length</span>(<span class="kw"><a href="../reference/ids.html">ids</a></span>(presults)) <span class="op">==</span><span class="st"> </span><span class="kw"><a href="../reference/count.html">count</a></span>(pquery)</a></code></pre></div>
<pre><code>## [1] TRUE</code></pre>
<p>Extracting subsets of results or manipulating the results into a more conventional R data structure is not easily generalizable. However, the <a href="https://github.com/hadley/purrr">purrr</a>, <a href="https://renkun.me/rlist/">rlist</a>, and <a href="https://cran.r-project.org/web/packages/data.tree/vignettes/data.tree.html">data.tree</a> packages are all potentially of interest for manipulating complex, nested list structures. For viewing the results in an interactive viewer, consider the <a href="https://github.com/timelyportfolio/listviewer">listviewer</a> package.</p>
</div>
<div id="fields-and-values" class="section level3">
<h3 class="hasAnchor">
<a href="#fields-and-values" class="anchor"></a>Fields and Values</h3>
<p><a href="https://docs.gdc.cancer.gov/API/Users_Guide/Search_and_Retrieval/#fields">[ GDC <code>fields</code> documentation ]</a></p>
<p>Central to querying and retrieving data from the GDC is the ability to specify which fields to return, filtering by fields and values, and faceting or aggregating. The GenomicDataCommons package includes two simple functions, <code><a href="../reference/available_fields.html">available_fields()</a></code> and <code><a href="../reference/default_fields.html">default_fields()</a></code>. Each can operate on a character(1) endpoint name (“cases”, “files”, “annotations”, or “projects”) or a <code>GDCQuery</code> object.</p>
<div class="sourceCode" id="cb25"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb25-1" data-line-number="1"><span class="kw"><a href="../reference/default_fields.html">default_fields</a></span>(<span class="st">'files'</span>)</a></code></pre></div>
<pre><code>##  [1] "access"                "acl"                  
##  [3] "created_datetime"      "data_category"        
##  [5] "data_format"           "data_type"            
##  [7] "error_type"            "experimental_strategy"
##  [9] "file_autocomplete"     "file_id"              
## [11] "file_name"             "file_size"            
## [13] "imaging_date"          "magnification"        
## [15] "md5sum"                "origin"               
## [17] "platform"              "read_pair_number"     
## [19] "revision"              "state"                
## [21] "state_comment"         "submitter_id"         
## [23] "tags"                  "type"                 
## [25] "updated_datetime"</code></pre>
<div class="sourceCode" id="cb27"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb27-1" data-line-number="1"><span class="co"># The number of fields available for files endpoint</span></a>
<a class="sourceLine" id="cb27-2" data-line-number="2"><span class="kw">length</span>(<span class="kw"><a href="../reference/available_fields.html">available_fields</a></span>(<span class="st">'files'</span>))</a></code></pre></div>
<pre><code>## [1] 651</code></pre>
<div class="sourceCode" id="cb29"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb29-1" data-line-number="1"><span class="co"># The first few fields available for files endpoint</span></a>
<a class="sourceLine" id="cb29-2" data-line-number="2"><span class="kw">head</span>(<span class="kw"><a href="../reference/available_fields.html">available_fields</a></span>(<span class="st">'files'</span>))</a></code></pre></div>
<pre><code>## [1] "access"                      "acl"                        
## [3] "analysis.analysis_id"        "analysis.analysis_type"     
## [5] "analysis.created_datetime"   "analysis.input_files.access"</code></pre>
<p>The fields to be returned by a query can be specified following a similar paradigm to that of the dplyr package. The <code><a href="../reference/select.html">select()</a></code> function is a verb that resets the fields slot of a <code>GDCQuery</code>; note that this is not quite analogous to the dplyr <code><a href="../reference/select.html">select()</a></code> verb that limits from already-present fields. We <em>completely replace</em> the fields when using <code><a href="../reference/select.html">select()</a></code> on a <code>GDCQuery</code>.</p>
<div class="sourceCode" id="cb31"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb31-1" data-line-number="1"><span class="co"># Default fields here</span></a>
<a class="sourceLine" id="cb31-2" data-line-number="2">qcases =<span class="st"> </span><span class="kw"><a href="../reference/query.html">cases</a></span>()</a>
<a class="sourceLine" id="cb31-3" data-line-number="3">qcases<span class="op">$</span>fields</a></code></pre></div>
<pre><code>##  [1] "aliquot_ids"              "analyte_ids"             
##  [3] "case_autocomplete"        "case_id"                 
##  [5] "created_datetime"         "days_to_index"           
##  [7] "days_to_lost_to_followup" "disease_type"            
##  [9] "index_date"               "lost_to_followup"        
## [11] "portion_ids"              "primary_site"            
## [13] "sample_ids"               "slide_ids"               
## [15] "state"                    "submitter_aliquot_ids"   
## [17] "submitter_analyte_ids"    "submitter_id"            
## [19] "submitter_portion_ids"    "submitter_sample_ids"    
## [21] "submitter_slide_ids"      "updated_datetime"</code></pre>
<div class="sourceCode" id="cb33"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb33-1" data-line-number="1"><span class="co"># set up query to use ALL available fields</span></a>
<a class="sourceLine" id="cb33-2" data-line-number="2"><span class="co"># Note that checking of fields is done by select()</span></a>
<a class="sourceLine" id="cb33-3" data-line-number="3">qcases =<span class="st"> </span><span class="kw"><a href="../reference/query.html">cases</a></span>() <span class="op">%&gt;%</span><span class="st"> </span>GenomicDataCommons<span class="op">::</span><span class="kw"><a href="http://www.rdocumentation.org/packages/GenomicDataCommons/topics/select">select</a></span>(<span class="kw"><a href="../reference/available_fields.html">available_fields</a></span>(<span class="st">'cases'</span>))</a>
<a class="sourceLine" id="cb33-4" data-line-number="4"><span class="kw">head</span>(qcases<span class="op">$</span>fields)</a></code></pre></div>
<pre><code>## [1] "case_id"                       "aliquot_ids"                  
## [3] "analyte_ids"                   "annotations.annotation_id"    
## [5] "annotations.case_id"           "annotations.case_submitter_id"</code></pre>
<p>Finding fields of interest is such a common operation that the GenomicDataCommons includes the <code><a href="../reference/grep_fields.html">grep_fields()</a></code> function and the <code>field_picker()</code> widget. See the appropriate help pages for details.</p>
</div>
<div id="facets-and-aggregation" class="section level3">
<h3 class="hasAnchor">
<a href="#facets-and-aggregation" class="anchor"></a>Facets and aggregation</h3>
<p><a href="https://docs.gdc.cancer.gov/API/Users_Guide/Search_and_Retrieval/#facets">[ GDC <code>facet</code> documentation ]</a></p>
<p>The GDC API offers a feature known as aggregation or faceting. By specifying one or more fields (of appropriate type), the GDC can return to us a count of the number of records matching each potential value. This is similar to the R <code>table</code> method. Multiple fields can be returned at once, but the GDC API does not have a cross-tabulation feature; all aggregations are only on one field at a time. Results of <code>aggregation()</code> calls come back as a list of data.frames (actually, tibbles).</p>
<div class="sourceCode" id="cb35"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb35-1" data-line-number="1"><span class="co"># total number of files of a specific type</span></a>
<a class="sourceLine" id="cb35-2" data-line-number="2">res =<span class="st"> </span><span class="kw"><a href="../reference/query.html">files</a></span>() <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="../reference/faceting.html">facet</a></span>(<span class="kw">c</span>(<span class="st">'type'</span>,<span class="st">'data_type'</span>)) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="../reference/aggregations.html">aggregations</a></span>()</a>
<a class="sourceLine" id="cb35-3" data-line-number="3">res<span class="op">$</span>type</a></code></pre></div>
<pre><code>##                            key doc_count
## 1      simple_somatic_mutation     64015
## 2   annotated_somatic_mutation     63580
## 3                aligned_reads     45985
## 4          copy_number_segment     44752
## 5              gene_expression     34713
## 6                  slide_image     30036
## 7       biospecimen_supplement     25151
## 8             mirna_expression     22976
## 9          clinical_supplement     12496
## 10      methylation_beta_value     12359
## 11 aggregated_somatic_mutation       186
## 12     masked_somatic_mutation       132</code></pre>
<p>Using <code><a href="../reference/aggregations.html">aggregations()</a></code> is an also easy way to learn the contents of individual fields and forms the basis for faceted search pages.</p>
</div>
<div id="filtering" class="section level3">
<h3 class="hasAnchor">
<a href="#filtering" class="anchor"></a>Filtering</h3>
<p><a href="https://docs.gdc.cancer.gov/API/Users_Guide/Search_and_Retrieval/#filters-specifying-the-query">[ GDC <code>filtering</code> documentation ]</a></p>
<p>The GenomicDataCommons package uses a form of non-standard evaluation to specify R-like queries that are then translated into an R list. That R list is, upon calling a method that fetches results from the GDC API, translated into the appropriate JSON string. The R expression uses the formula interface as suggested by Hadley Wickham in his <a href="https://cran.r-project.org/web/packages/dplyr/vignettes/nse.html">vignette on non-standard evaluation</a></p>
<blockquote>
<p>It’s best to use a formula because a formula captures both the expression to evaluate and the environment where the evaluation occurs. This is important if the expression is a mixture of variables in a data frame and objects in the local environment [for example].</p>
</blockquote>
<p>For the user, these details will not be too important except to note that a filter expression must begin with a “~”.</p>
<div class="sourceCode" id="cb37"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb37-1" data-line-number="1">qfiles =<span class="st"> </span><span class="kw"><a href="../reference/query.html">files</a></span>()</a>
<a class="sourceLine" id="cb37-2" data-line-number="2">qfiles <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="../reference/count.html">count</a></span>() <span class="co"># all files</span></a></code></pre></div>
<pre><code>## [1] 356381</code></pre>
<p>To limit the file type, we can refer back to the <a href="#facets-and-aggregation">section on faceting</a> to see the possible values for the file field “type”. For example, to filter file results to only “gene_expression” files, we simply specify a filter.</p>
<div class="sourceCode" id="cb39"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb39-1" data-line-number="1">qfiles =<span class="st"> </span><span class="kw"><a href="../reference/query.html">files</a></span>() <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="../reference/filtering.html">filter</a></span>(<span class="op">~</span><span class="st"> </span>type <span class="op">==</span><span class="st"> 'gene_expression'</span>)</a>
<a class="sourceLine" id="cb39-2" data-line-number="2"><span class="co"># here is what the filter looks like after translation</span></a>
<a class="sourceLine" id="cb39-3" data-line-number="3"><span class="kw">str</span>(<span class="kw"><a href="../reference/filtering.html">get_filter</a></span>(qfiles))</a></code></pre></div>
<pre><code>## List of 2
##  $ op     : 'scalar' chr "="
##  $ content:List of 2
##   ..$ field: chr "type"
##   ..$ value: chr "gene_expression"</code></pre>
<p>What if we want to create a filter based on the project (‘TCGA-OVCA’, for example)? Well, we have a couple of possible ways to discover available fields. The first is based on base R functionality and some intuition.</p>
<div class="sourceCode" id="cb41"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb41-1" data-line-number="1"><span class="kw">grep</span>(<span class="st">'pro'</span>,<span class="kw"><a href="../reference/available_fields.html">available_fields</a></span>(<span class="st">'files'</span>),<span class="dt">value=</span><span class="ot">TRUE</span>)</a></code></pre></div>
<pre><code>##  [1] "cases.diagnoses.progression_free_survival"               
##  [2] "cases.diagnoses.progression_free_survival_event"         
##  [3] "cases.diagnoses.progression_or_recurrence"               
##  [4] "cases.project.dbgap_accession_number"                    
##  [5] "cases.project.disease_type"                              
##  [6] "cases.project.intended_release_date"                     
##  [7] "cases.project.name"                                      
##  [8] "cases.project.primary_site"                              
##  [9] "cases.project.program.dbgap_accession_number"            
## [10] "cases.project.program.name"                              
## [11] "cases.project.program.program_id"                        
## [12] "cases.project.project_id"                                
## [13] "cases.project.releasable"                                
## [14] "cases.project.released"                                  
## [15] "cases.project.state"                                     
## [16] "cases.samples.days_to_sample_procurement"                
## [17] "cases.samples.method_of_sample_procurement"              
## [18] "cases.samples.portions.slides.number_proliferating_cells"
## [19] "cases.tissue_source_site.project"</code></pre>
<p>Interestingly, the project information is “nested” inside the case. We don’t need to know that detail other than to know that we now have a few potential guesses for where our information might be in the files records. We need to know where because we need to construct the appropriate filter.</p>
<div class="sourceCode" id="cb43"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb43-1" data-line-number="1"><span class="kw"><a href="../reference/query.html">files</a></span>() <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="../reference/faceting.html">facet</a></span>(<span class="st">'cases.project.project_id'</span>) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="../reference/aggregations.html">aggregations</a></span>()</a></code></pre></div>
<pre><code>## $cases.project.project_id
##            key doc_count
## 1        FM-AD     36134
## 2    TCGA-BRCA     31511
## 3    TCGA-LUAD     17051
## 4    TCGA-UCEC     16130
## 5    TCGA-HNSC     15266
## 6      TCGA-OV     15057
## 7    TCGA-THCA     14420
## 8    TCGA-LUSC     15323
## 9     TCGA-LGG     14723
## 10   TCGA-KIRC     15082
## 11   TCGA-PRAD     14287
## 12   TCGA-COAD     14270
## 13    TCGA-GBM     11973
## 14   TCGA-SKCM     12724
## 15   TCGA-STAD     12845
## 16   TCGA-BLCA     11710
## 17   TCGA-LIHC     10814
## 18   TCGA-CESC      8593
## 19   TCGA-KIRP      8506
## 20   TCGA-SARC      7493
## 21   TCGA-PAAD      5306
## 22   TCGA-ESCA      5270
## 23   TCGA-PCPG      5032
## 24   TCGA-READ      4918
## 25   TCGA-TGCT      4217
## 26   TCGA-THYM      3444
## 27   TCGA-LAML      3960
## 28  TARGET-NBL      2795
## 29    TCGA-ACC      2546
## 30   TCGA-KICH      2324
## 31   TCGA-MESO      2330
## 32  TARGET-AML      2170
## 33    TCGA-UVM      2179
## 34    TCGA-UCS      1658
## 35   TARGET-WT      1406
## 36   TCGA-DLBC      1330
## 37   TCGA-CHOL      1348
## 38   TARGET-OS        47
## 39   TARGET-RT       174
## 40 TARGET-CCSK        15</code></pre>
<p>We note that <code>cases.project.project_id</code> looks like it is a good fit. We also note that <code>TCGA-OV</code> is the correct project_id, not <code>TCGA-OVCA</code>. Note that <em>unlike with dplyr and friends, the <code><a href="../reference/filtering.html">filter()</a></code> method here <strong>replaces</strong> the filter and does not build on any previous filters</em>.</p>
<div class="sourceCode" id="cb45"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb45-1" data-line-number="1">qfiles =<span class="st"> </span><span class="kw"><a href="../reference/query.html">files</a></span>() <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb45-2" data-line-number="2"><span class="st">    </span><span class="kw"><a href="../reference/filtering.html">filter</a></span>( <span class="op">~</span><span class="st"> </span>cases.project.project_id <span class="op">==</span><span class="st"> 'TCGA-OV'</span> <span class="op">&amp;</span><span class="st"> </span>type <span class="op">==</span><span class="st"> 'gene_expression'</span>)</a>
<a class="sourceLine" id="cb45-3" data-line-number="3"><span class="kw">str</span>(<span class="kw"><a href="../reference/filtering.html">get_filter</a></span>(qfiles))</a></code></pre></div>
<pre><code>## List of 2
##  $ op     : 'scalar' chr "and"
##  $ content:List of 2
##   ..$ :List of 2
##   .. ..$ op     : 'scalar' chr "="
##   .. ..$ content:List of 2
##   .. .. ..$ field: chr "cases.project.project_id"
##   .. .. ..$ value: chr "TCGA-OV"
##   ..$ :List of 2
##   .. ..$ op     : 'scalar' chr "="
##   .. ..$ content:List of 2
##   .. .. ..$ field: chr "type"
##   .. .. ..$ value: chr "gene_expression"</code></pre>
<div class="sourceCode" id="cb47"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb47-1" data-line-number="1">qfiles <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="../reference/count.html">count</a></span>()</a></code></pre></div>
<pre><code>## [1] 1137</code></pre>
<p>Asking for a <code><a href="../reference/count.html">count()</a></code> of results given these new filter criteria gives <code>r qfiles %&gt;% count()</code> results. Generating a manifest for bulk downloads is as simple as asking for the manifest from the current query.</p>
<div class="sourceCode" id="cb49"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb49-1" data-line-number="1">manifest_df =<span class="st"> </span>qfiles <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="../reference/manifest.html">manifest</a></span>()</a>
<a class="sourceLine" id="cb49-2" data-line-number="2"><span class="kw">head</span>(manifest_df)</a></code></pre></div>
<pre><code>## # A tibble: 6 x 5
##   id                filename                  md5               size state
##   &lt;chr&gt;             &lt;chr&gt;                     &lt;chr&gt;            &lt;int&gt; &lt;chr&gt;
## 1 e0e2ad9d-357f-4f… 67bee95c-2a85-443b-b662-… 56cf43078ba1e2… 262733 live 
## 2 b84b9118-3a5a-41… 94823ae2-8fb8-4be7-b2ca-… 47cb1de12d28c9… 256377 live 
## 3 1af12665-932f-41… 0ffb5d54-6522-4074-9c41-… c26f786d701401… 557285 live 
## 4 04a40917-633c-48… 97d62176-1e34-4508-886b-… 9be069378aed12… 255779 live 
## 5 b7c1da28-59e1-48… ab8603dd-2f94-4c83-9927-… 4b5a4698c099a7… 249380 live 
## 6 560d56bc-35ee-43… 30b1f3cd-604d-4a13-a196-… b58bbd636b6938… 261497 live</code></pre>
<p>Note that we might still not be quite there. Looking at filenames, there are suspiciously named files that might include “FPKM”, “FPKM-UQ”, or “counts”. Another round of <code>grep</code> and <code>available_fields</code>, looking for “type” turned up that the field “analysis.workflow_type” has the appropriate filter criteria.</p>
<div class="sourceCode" id="cb51"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb51-1" data-line-number="1">qfiles =<span class="st"> </span><span class="kw"><a href="../reference/query.html">files</a></span>() <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="../reference/filtering.html">filter</a></span>( <span class="op">~</span><span class="st"> </span>cases.project.project_id <span class="op">==</span><span class="st"> 'TCGA-OV'</span> <span class="op">&amp;</span></a>
<a class="sourceLine" id="cb51-2" data-line-number="2"><span class="st">                            </span>type <span class="op">==</span><span class="st"> 'gene_expression'</span> <span class="op">&amp;</span></a>
<a class="sourceLine" id="cb51-3" data-line-number="3"><span class="st">                            </span>analysis.workflow_type <span class="op">==</span><span class="st"> 'HTSeq - Counts'</span>)</a>
<a class="sourceLine" id="cb51-4" data-line-number="4">manifest_df =<span class="st"> </span>qfiles <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="../reference/manifest.html">manifest</a></span>()</a>
<a class="sourceLine" id="cb51-5" data-line-number="5"><span class="kw">nrow</span>(manifest_df)</a></code></pre></div>
<pre><code>## [1] 379</code></pre>
<p>The GDC Data Transfer Tool can be used (from R, <code><a href="../reference/transfer.html">transfer()</a></code> or from the command-line) to orchestrate high-performance, restartable transfers of all the files in the manifest. See <a href="bulk-downloads">the bulk downloads section</a> for details.</p>
</div>
</div>
<div id="authentication" class="section level2">
<h2 class="hasAnchor">
<a href="#authentication" class="anchor"></a>Authentication</h2>
<p><a href="https://docs.gdc.cancer.gov/API/Users_Guide/Search_and_Retrieval/#facets">[ GDC authentication documentation ]</a></p>
<p>The GDC offers both “controlled-access” and “open” data. As of this writing, only data stored as files is “controlled-access”; that is, metadata accessible via the GDC is all “open” data and some files are “open” and some are “controlled-access”. Controlled-access data are only available after <a href="https://gdc.cancer.gov/access-data/obtaining-access-controlled-data">going through the process of obtaining access.</a></p>
<p>After controlled-access to one or more datasets has been granted, logging into the GDC web portal will allow you to <a href="https://docs.gdc.cancer.gov/Data_Portal/Users_Guide/Authentication/#gdc-authentication-tokens">access a GDC authentication token</a>, which can be downloaded and then used to access available controlled-access data via the GenomicDataCommons package.</p>
<p>The GenomicDataCommons uses authentication tokens only for downloading data (see <code>transfer</code> and <code>gdcdata</code> documentation). The package includes a helper function, <code>gdc_token</code>, that looks for the token to be stored in one of three ways (resolved in this order):</p>
<ol style="list-style-type: decimal">
<li>As a string stored in the environment variable, <code>GDC_TOKEN</code>
</li>
<li>As a file, stored in the file named by the environment variable, <code>GDC_TOKEN_FILE</code>
</li>
<li>In a file in the user home directory, called <code>.gdc_token</code>
</li>
</ol>
<p>As a concrete example:</p>
<div class="sourceCode" id="cb53"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb53-1" data-line-number="1">token =<span class="st"> </span><span class="kw"><a href="../reference/gdc_token.html">gdc_token</a></span>()</a>
<a class="sourceLine" id="cb53-2" data-line-number="2"><span class="kw"><a href="../reference/transfer.html">transfer</a></span>(...,<span class="dt">token=</span>token)</a>
<a class="sourceLine" id="cb53-3" data-line-number="3"><span class="co"># or</span></a>
<a class="sourceLine" id="cb53-4" data-line-number="4"><span class="kw"><a href="../reference/transfer.html">transfer</a></span>(...,<span class="dt">token=</span><span class="kw">get_token</span>())</a></code></pre></div>
</div>
<div id="datafile-access-and-download" class="section level2">
<h2 class="hasAnchor">
<a href="#datafile-access-and-download" class="anchor"></a>Datafile access and download</h2>
<div id="data-downloads-via-the-gdc-api" class="section level3">
<h3 class="hasAnchor">
<a href="#data-downloads-via-the-gdc-api" class="anchor"></a>Data downloads via the GDC API</h3>
<p>The <code>gdcdata</code> function takes a character vector of one or more file ids. A simple way of producing such a vector is to produce a <code>manifest</code> data frame and then pass in the first column, which will contain file ids.</p>
<div class="sourceCode" id="cb54"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb54-1" data-line-number="1">fnames =<span class="st"> </span><span class="kw"><a href="../reference/gdcdata.html">gdcdata</a></span>(manifest_df<span class="op">$</span>id[<span class="dv">1</span><span class="op">:</span><span class="dv">2</span>],<span class="dt">progress=</span><span class="ot">FALSE</span>)</a></code></pre></div>
<p>Note that for controlled-access data, a GDC <a href="#authentication">authentication token</a> is required. Using the <code>BiocParallel</code> package may be useful for downloading in parallel, particularly for large numbers of smallish files.</p>
</div>
<div id="bulk-downloads" class="section level3">
<h3 class="hasAnchor">
<a href="#bulk-downloads" class="anchor"></a>Bulk downloads</h3>
<p>The bulk download functionality is only efficient (as of v1.2.0 of the GDC Data Transfer Tool) for relatively large files, so use this approach only when transferring BAM files or larger VCF files, for example. Otherwise, consider using the approach shown above, perhaps in parallel.</p>
<div class="sourceCode" id="cb55"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb55-1" data-line-number="1">fnames =<span class="st"> </span><span class="kw"><a href="../reference/gdcdata.html">gdcdata</a></span>(manifest_df<span class="op">$</span>id[<span class="dv">3</span><span class="op">:</span><span class="dv">10</span>], <span class="dt">access_method =</span> <span class="st">'client'</span>)</a></code></pre></div>
</div>
<div id="bam-slicing" class="section level3">
<h3 class="hasAnchor">
<a href="#bam-slicing" class="anchor"></a>BAM slicing</h3>
</div>
</div>
</div>
<div id="use-cases" class="section level1">
<h1 class="hasAnchor">
<a href="#use-cases" class="anchor"></a>Use Cases</h1>
<div id="cases" class="section level2">
<h2 class="hasAnchor">
<a href="#cases" class="anchor"></a>Cases</h2>
<div id="how-many-cases-are-there-per-project_id" class="section level3">
<h3 class="hasAnchor">
<a href="#how-many-cases-are-there-per-project_id" class="anchor"></a>How many cases are there per project_id?</h3>
<div class="sourceCode" id="cb56"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb56-1" data-line-number="1">res =<span class="st"> </span><span class="kw"><a href="../reference/query.html">cases</a></span>() <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="../reference/faceting.html">facet</a></span>(<span class="st">"project.project_id"</span>) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="../reference/aggregations.html">aggregations</a></span>()</a>
<a class="sourceLine" id="cb56-2" data-line-number="2"><span class="kw">head</span>(res)</a></code></pre></div>
<pre><code>## $project.project_id
##            key doc_count
## 1        FM-AD     18004
## 2   TARGET-NBL      1127
## 3    TCGA-BRCA      1098
## 4   TARGET-AML       988
## 5    TARGET-WT       652
## 6     TCGA-GBM       617
## 7      TCGA-OV       608
## 8    TCGA-LUAD       585
## 9    TCGA-UCEC       560
## 10   TCGA-KIRC       537
## 11   TCGA-HNSC       528
## 12    TCGA-LGG       516
## 13   TCGA-THCA       507
## 14   TCGA-LUSC       504
## 15   TCGA-PRAD       500
## 16   TCGA-SKCM       470
## 17   TCGA-COAD       461
## 18   TCGA-STAD       443
## 19   TCGA-BLCA       412
## 20   TARGET-OS       381
## 21   TCGA-LIHC       377
## 22   TCGA-CESC       307
## 23   TCGA-KIRP       291
## 24   TCGA-SARC       261
## 25   TCGA-LAML       200
## 26   TCGA-ESCA       185
## 27   TCGA-PAAD       185
## 28   TCGA-PCPG       179
## 29   TCGA-READ       172
## 30   TCGA-TGCT       150
## 31   TCGA-THYM       124
## 32   TCGA-KICH       113
## 33    TCGA-ACC        92
## 34   TCGA-MESO        87
## 35    TCGA-UVM        80
## 36   TARGET-RT        75
## 37   TCGA-DLBC        58
## 38    TCGA-UCS        57
## 39   TCGA-CHOL        51
## 40 TARGET-CCSK        13</code></pre>
<div class="sourceCode" id="cb58"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb58-1" data-line-number="1"><span class="kw">library</span>(ggplot2)</a>
<a class="sourceLine" id="cb58-2" data-line-number="2"><span class="kw"><a href="http://www.rdocumentation.org/packages/ggplot2/topics/ggplot">ggplot</a></span>(res<span class="op">$</span>project.project_id,<span class="kw"><a href="http://www.rdocumentation.org/packages/ggplot2/topics/aes">aes</a></span>(<span class="dt">x =</span> key, <span class="dt">y =</span> doc_count)) <span class="op">+</span></a>
<a class="sourceLine" id="cb58-3" data-line-number="3"><span class="st">    </span><span class="kw"><a href="http://www.rdocumentation.org/packages/ggplot2/topics/geom_bar">geom_bar</a></span>(<span class="dt">stat=</span><span class="st">'identity'</span>) <span class="op">+</span></a>
<a class="sourceLine" id="cb58-4" data-line-number="4"><span class="st">    </span><span class="kw"><a href="http://www.rdocumentation.org/packages/ggplot2/topics/theme">theme</a></span>(<span class="dt">axis.text.x =</span> <span class="kw"><a href="http://www.rdocumentation.org/packages/ggplot2/topics/element">element_text</a></span>(<span class="dt">angle =</span> <span class="dv">45</span>, <span class="dt">hjust =</span> <span class="dv">1</span>))</a></code></pre></div>
<p><img src="overview_files/figure-html/casesPerProject-1.png" width="700"></p>
</div>
<div id="how-many-cases-are-included-in-all-target-projects" class="section level3">
<h3 class="hasAnchor">
<a href="#how-many-cases-are-included-in-all-target-projects" class="anchor"></a>How many cases are included in all TARGET projects?</h3>
<div class="sourceCode" id="cb59"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb59-1" data-line-number="1"><span class="kw"><a href="../reference/query.html">cases</a></span>() <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="../reference/filtering.html">filter</a></span>(<span class="op">~</span><span class="st"> </span>project.program.name<span class="op">==</span><span class="st">'TARGET'</span>) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="../reference/count.html">count</a></span>()</a></code></pre></div>
<pre><code>## [1] 3236</code></pre>
</div>
<div id="how-many-cases-are-included-in-all-tcga-projects" class="section level3">
<h3 class="hasAnchor">
<a href="#how-many-cases-are-included-in-all-tcga-projects" class="anchor"></a>How many cases are included in all TCGA projects?</h3>
<div class="sourceCode" id="cb61"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb61-1" data-line-number="1"><span class="kw"><a href="../reference/query.html">cases</a></span>() <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="../reference/filtering.html">filter</a></span>(<span class="op">~</span><span class="st"> </span>project.program.name<span class="op">==</span><span class="st">'TCGA'</span>) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="../reference/count.html">count</a></span>()</a></code></pre></div>
<pre><code>## [1] 11315</code></pre>
</div>
<div id="what-is-the-breakdown-of-sample-types-in-tcga-brca" class="section level3">
<h3 class="hasAnchor">
<a href="#what-is-the-breakdown-of-sample-types-in-tcga-brca" class="anchor"></a>What is the breakdown of sample types in TCGA-BRCA?</h3>
<div class="sourceCode" id="cb63"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb63-1" data-line-number="1"><span class="co"># The need to do the "&amp;" here is a requirement of the</span></a>
<a class="sourceLine" id="cb63-2" data-line-number="2"><span class="co"># current version of the GDC API. I have filed a feature</span></a>
<a class="sourceLine" id="cb63-3" data-line-number="3"><span class="co"># request to remove this requirement.</span></a>
<a class="sourceLine" id="cb63-4" data-line-number="4">resp =<span class="st"> </span><span class="kw"><a href="../reference/query.html">cases</a></span>() <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="../reference/filtering.html">filter</a></span>(<span class="op">~</span><span class="st"> </span>project.project_id<span class="op">==</span><span class="st">'TCGA-BRCA'</span> <span class="op">&amp;</span></a>
<a class="sourceLine" id="cb63-5" data-line-number="5"><span class="st">                              </span>project.project_id<span class="op">==</span><span class="st">'TCGA-BRCA'</span> ) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb63-6" data-line-number="6"><span class="st">    </span><span class="kw"><a href="../reference/faceting.html">facet</a></span>(<span class="st">'samples.sample_type'</span>) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="../reference/aggregations.html">aggregations</a></span>()</a>
<a class="sourceLine" id="cb63-7" data-line-number="7">resp<span class="op">$</span>samples.sample_type</a></code></pre></div>
<pre><code>##                    key doc_count
## 1        Primary Tumor      1098
## 2 Blood Derived Normal      1011
## 3  Solid Tissue Normal       162
## 4           Metastatic         7</code></pre>
</div>
<div id="fetch-all-samples-in-tcga-brca-that-use-solid-tissue-as-a-normal-" class="section level3">
<h3 class="hasAnchor">
<a href="#fetch-all-samples-in-tcga-brca-that-use-solid-tissue-as-a-normal-" class="anchor"></a>Fetch all samples in TCGA-BRCA that use “Solid Tissue” as a normal.</h3>
<div class="sourceCode" id="cb65"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb65-1" data-line-number="1"><span class="co"># The need to do the "&amp;" here is a requirement of the</span></a>
<a class="sourceLine" id="cb65-2" data-line-number="2"><span class="co"># current version of the GDC API. I have filed a feature</span></a>
<a class="sourceLine" id="cb65-3" data-line-number="3"><span class="co"># request to remove this requirement.</span></a>
<a class="sourceLine" id="cb65-4" data-line-number="4">resp =<span class="st"> </span><span class="kw"><a href="../reference/query.html">cases</a></span>() <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="../reference/filtering.html">filter</a></span>(<span class="op">~</span><span class="st"> </span>project.project_id<span class="op">==</span><span class="st">'TCGA-BRCA'</span> <span class="op">&amp;</span></a>
<a class="sourceLine" id="cb65-5" data-line-number="5"><span class="st">                              </span>samples.sample_type<span class="op">==</span><span class="st">'Solid Tissue Normal'</span>) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb65-6" data-line-number="6"><span class="st">    </span>GenomicDataCommons<span class="op">::</span><span class="kw"><a href="http://www.rdocumentation.org/packages/GenomicDataCommons/topics/select">select</a></span>(<span class="kw">c</span>(<span class="kw"><a href="../reference/default_fields.html">default_fields</a></span>(<span class="kw"><a href="../reference/query.html">cases</a></span>()),<span class="st">'samples.sample_type'</span>)) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb65-7" data-line-number="7"><span class="st">    </span><span class="kw"><a href="../reference/response.html">response_all</a></span>()</a>
<a class="sourceLine" id="cb65-8" data-line-number="8"><span class="kw"><a href="../reference/count.html">count</a></span>(resp)</a></code></pre></div>
<pre><code>## [1] 162</code></pre>
<div class="sourceCode" id="cb67"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb67-1" data-line-number="1">res =<span class="st"> </span>resp <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="../reference/results.html">results</a></span>()</a>
<a class="sourceLine" id="cb67-2" data-line-number="2"><span class="kw">str</span>(res[<span class="dv">1</span>],<span class="dt">list.len=</span><span class="dv">6</span>)</a></code></pre></div>
<pre><code>## List of 1
##  $ updated_datetime: chr [1:162] "2018-05-21T16:07:40.645885-05:00" "2018-05-21T16:07:40.645885-05:00" "2018-05-21T16:07:40.645885-05:00" "2018-05-21T16:07:40.645885-05:00" ...</code></pre>
<div class="sourceCode" id="cb69"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb69-1" data-line-number="1"><span class="kw">head</span>(<span class="kw"><a href="../reference/ids.html">ids</a></span>(resp))</a></code></pre></div>
<pre><code>## [1] "bfd49783-1767-469b-9d79-7822301c5efc"
## [2] "a0576d44-9add-429b-9eeb-f7b234b7385f"
## [3] "955d4263-61f7-42e8-8a6e-772a0d6c209d"
## [4] "3d676bba-154b-4d22-ab59-d4d4da051b94"
## [5] "e666081b-caaf-4e8b-9446-807be71201a5"
## [6] "5fe3c51f-65e2-4da6-8924-a6fae17c022d"</code></pre>
</div>
</div>
<div id="files" class="section level2">
<h2 class="hasAnchor">
<a href="#files" class="anchor"></a>Files</h2>
<div id="how-many-of-each-type-of-file-are-available" class="section level3">
<h3 class="hasAnchor">
<a href="#how-many-of-each-type-of-file-are-available" class="anchor"></a>How many of each type of file are available?</h3>
<div class="sourceCode" id="cb71"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb71-1" data-line-number="1">res =<span class="st"> </span><span class="kw"><a href="../reference/query.html">files</a></span>() <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="../reference/faceting.html">facet</a></span>(<span class="st">'type'</span>) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="../reference/aggregations.html">aggregations</a></span>()</a>
<a class="sourceLine" id="cb71-2" data-line-number="2">res<span class="op">$</span>type</a></code></pre></div>
<pre><code>##                            key doc_count
## 1      simple_somatic_mutation     64015
## 2   annotated_somatic_mutation     63580
## 3                aligned_reads     45985
## 4          copy_number_segment     44752
## 5              gene_expression     34713
## 6                  slide_image     30036
## 7       biospecimen_supplement     25151
## 8             mirna_expression     22976
## 9          clinical_supplement     12496
## 10      methylation_beta_value     12359
## 11 aggregated_somatic_mutation       186
## 12     masked_somatic_mutation       132</code></pre>
<div class="sourceCode" id="cb73"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb73-1" data-line-number="1"><span class="kw"><a href="http://www.rdocumentation.org/packages/ggplot2/topics/ggplot">ggplot</a></span>(res<span class="op">$</span>type,<span class="kw"><a href="http://www.rdocumentation.org/packages/ggplot2/topics/aes">aes</a></span>(<span class="dt">x =</span> key,<span class="dt">y =</span> doc_count)) <span class="op">+</span><span class="st"> </span><span class="kw"><a href="http://www.rdocumentation.org/packages/ggplot2/topics/geom_bar">geom_bar</a></span>(<span class="dt">stat=</span><span class="st">'identity'</span>) <span class="op">+</span></a>
<a class="sourceLine" id="cb73-2" data-line-number="2"><span class="st">    </span><span class="kw"><a href="http://www.rdocumentation.org/packages/ggplot2/topics/theme">theme</a></span>(<span class="dt">axis.text.x =</span> <span class="kw"><a href="http://www.rdocumentation.org/packages/ggplot2/topics/element">element_text</a></span>(<span class="dt">angle =</span> <span class="dv">45</span>, <span class="dt">hjust =</span> <span class="dv">1</span>))</a></code></pre></div>
<p><img src="overview_files/figure-html/filesVCFCount-1.png" width="700"></p>
</div>
<div id="find-gene-level-rna-seq-quantification-files-for-gbm" class="section level3">
<h3 class="hasAnchor">
<a href="#find-gene-level-rna-seq-quantification-files-for-gbm" class="anchor"></a>Find gene-level RNA-seq quantification files for GBM</h3>
<div class="sourceCode" id="cb74"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb74-1" data-line-number="1">q =<span class="st"> </span><span class="kw"><a href="../reference/query.html">files</a></span>() <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb74-2" data-line-number="2"><span class="st">    </span>GenomicDataCommons<span class="op">::</span><span class="kw"><a href="http://www.rdocumentation.org/packages/GenomicDataCommons/topics/select">select</a></span>(<span class="kw"><a href="../reference/available_fields.html">available_fields</a></span>(<span class="st">'files'</span>)) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb74-3" data-line-number="3"><span class="st">    </span><span class="kw"><a href="../reference/filtering.html">filter</a></span>(<span class="op">~</span><span class="st"> </span>cases.project.project_id<span class="op">==</span><span class="st">'TCGA-GBM'</span> <span class="op">&amp;</span></a>
<a class="sourceLine" id="cb74-4" data-line-number="4"><span class="st">               </span>data_type<span class="op">==</span><span class="st">'Gene Expression Quantification'</span>)</a>
<a class="sourceLine" id="cb74-5" data-line-number="5">q <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="../reference/faceting.html">facet</a></span>(<span class="st">'analysis.workflow_type'</span>) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="../reference/aggregations.html">aggregations</a></span>()</a></code></pre></div>
<pre><code>## $analysis.workflow_type
##               key doc_count
## 1  HTSeq - Counts       174
## 2    HTSeq - FPKM       174
## 3 HTSeq - FPKM-UQ       174</code></pre>
<div class="sourceCode" id="cb76"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb76-1" data-line-number="1"><span class="co"># so need to add another filter</span></a>
<a class="sourceLine" id="cb76-2" data-line-number="2">file_ids =<span class="st"> </span>q <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="../reference/filtering.html">filter</a></span>(<span class="op">~</span><span class="st"> </span>cases.project.project_id<span class="op">==</span><span class="st">'TCGA-GBM'</span> <span class="op">&amp;</span></a>
<a class="sourceLine" id="cb76-3" data-line-number="3"><span class="st">                            </span>data_type<span class="op">==</span><span class="st">'Gene Expression Quantification'</span> <span class="op">&amp;</span></a>
<a class="sourceLine" id="cb76-4" data-line-number="4"><span class="st">                            </span>analysis.workflow_type <span class="op">==</span><span class="st"> 'HTSeq - Counts'</span>) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb76-5" data-line-number="5"><span class="st">    </span>GenomicDataCommons<span class="op">::</span><span class="kw"><a href="http://www.rdocumentation.org/packages/GenomicDataCommons/topics/select">select</a></span>(<span class="st">'file_id'</span>) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb76-6" data-line-number="6"><span class="st">    </span><span class="kw"><a href="../reference/response.html">response_all</a></span>() <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb76-7" data-line-number="7"><span class="st">    </span><span class="kw"><a href="../reference/ids.html">ids</a></span>()</a></code></pre></div>
</div>
</div>
<div id="slicing" class="section level2">
<h2 class="hasAnchor">
<a href="#slicing" class="anchor"></a>Slicing</h2>
<div id="get-all-bam-file-ids-from-tcga-gbm" class="section level3">
<h3 class="hasAnchor">
<a href="#get-all-bam-file-ids-from-tcga-gbm" class="anchor"></a>Get all BAM file ids from TCGA-GBM</h3>
<p><strong>I need to figure out how to do slicing reproducibly in a testing environment and for vignette building</strong>.</p>
<div class="sourceCode" id="cb77"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb77-1" data-line-number="1">q =<span class="st"> </span><span class="kw"><a href="../reference/query.html">files</a></span>() <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb77-2" data-line-number="2"><span class="st">    </span>GenomicDataCommons<span class="op">::</span><span class="kw"><a href="http://www.rdocumentation.org/packages/GenomicDataCommons/topics/select">select</a></span>(<span class="kw"><a href="../reference/available_fields.html">available_fields</a></span>(<span class="st">'files'</span>)) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb77-3" data-line-number="3"><span class="st">    </span><span class="kw"><a href="../reference/filtering.html">filter</a></span>(<span class="op">~</span><span class="st"> </span>cases.project.project_id <span class="op">==</span><span class="st"> 'TCGA-GBM'</span> <span class="op">&amp;</span></a>
<a class="sourceLine" id="cb77-4" data-line-number="4"><span class="st">               </span>data_type <span class="op">==</span><span class="st"> 'Aligned Reads'</span> <span class="op">&amp;</span></a>
<a class="sourceLine" id="cb77-5" data-line-number="5"><span class="st">               </span>experimental_strategy <span class="op">==</span><span class="st"> 'RNA-Seq'</span> <span class="op">&amp;</span></a>
<a class="sourceLine" id="cb77-6" data-line-number="6"><span class="st">               </span>data_format <span class="op">==</span><span class="st"> 'BAM'</span>)</a>
<a class="sourceLine" id="cb77-7" data-line-number="7">file_ids =<span class="st"> </span>q <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="../reference/response.html">response_all</a></span>() <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="../reference/ids.html">ids</a></span>()</a></code></pre></div>
<div class="sourceCode" id="cb78"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb78-1" data-line-number="1">bamfile =<span class="st"> </span><span class="kw"><a href="../reference/slicing.html">slicing</a></span>(file_ids[<span class="dv">1</span>],<span class="dt">regions=</span><span class="st">"chr12:6534405-6538375"</span>,<span class="dt">token=</span><span class="kw"><a href="../reference/gdc_token.html">gdc_token</a></span>())</a>
<a class="sourceLine" id="cb78-2" data-line-number="2"><span class="kw">library</span>(GenomicAlignments)</a>
<a class="sourceLine" id="cb78-3" data-line-number="3">aligns =<span class="st"> </span><span class="kw"><a href="http://www.rdocumentation.org/packages/GenomicAlignments/topics/readGAlignments">readGAlignments</a></span>(bamfile)</a></code></pre></div>
</div>
</div>
</div>
<div id="troubleshooting" class="section level1">
<h1 class="hasAnchor">
<a href="#troubleshooting" class="anchor"></a>Troubleshooting</h1>
<div id="ssl-connection-errors" class="section level2">
<h2 class="hasAnchor">
<a href="#ssl-connection-errors" class="anchor"></a>SSL connection errors</h2>
<ul>
<li>Symptom: Trying to connect to the API results in:</li>
</ul>
<pre><code>Error in curl::curl_fetch_memory(url, handle = handle) :
SSL connect error</code></pre>
<ul>
<li>Possible solutions: The <a href="http://stackoverflow.com/a/42599546/459633">issue is that the GDC supports only recent security Transport Layer Security (TLS)</a>, so the only known fix is to upgrade the system <code>openssl</code> to version 1.0.1 or later.
<ul>
<li>
<a href="https://github.com/Bioconductor/GenomicDataCommons/issues/35#issuecomment-284233510">[Mac OS]</a>,</li>
<li><a href="http://askubuntu.com/a/434245">[Ubuntu]</a></li>
<li>
<a href="https://www.liquidweb.com/kb/update-and-patch-openssl-for-the-ccs-injection-vulnerability/">[Centos/RHEL]</a>. After upgrading <code>openssl</code>, reinstall the R <code>curl</code> and <code>httr</code> packages.</li>
</ul>
</li>
</ul>
</div>
</div>
<div id="sessioninfo" class="section level1">
<h1 class="hasAnchor">
<a href="#sessioninfo" class="anchor"></a>sessionInfo()</h1>
<div class="sourceCode" id="cb80"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb80-1" data-line-number="1"><span class="kw">sessionInfo</span>()</a></code></pre></div>
<pre><code>## R version 3.5.1 (2018-07-02)
## Platform: x86_64-apple-darwin15.6.0 (64-bit)
## Running under: macOS Sierra 10.12.6
## 
## Matrix products: default
## BLAS: /Library/Frameworks/R.framework/Versions/3.5/Resources/lib/libRblas.0.dylib
## LAPACK: /Library/Frameworks/R.framework/Versions/3.5/Resources/lib/libRlapack.dylib
## 
## locale:
## [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8
## 
## attached base packages:
## [1] stats     graphics  grDevices utils     datasets  methods   base     
## 
## other attached packages:
## [1] ggplot2_3.0.0            GenomicDataCommons_1.4.1
## [3] magrittr_1.5             knitr_1.20              
## [5] BiocStyle_2.8.2         
## 
## loaded via a namespace (and not attached):
##  [1] Rcpp_0.12.18                lattice_0.20-35            
##  [3] assertthat_0.2.0            rprojroot_1.3-2            
##  [5] digest_0.6.16               utf8_1.1.4                 
##  [7] plyr_1.8.4                  R6_2.2.2                   
##  [9] GenomeInfoDb_1.16.0         backports_1.1.2            
## [11] stats4_3.5.1                evaluate_0.11              
## [13] httr_1.3.1                  pillar_1.3.0               
## [15] zlibbioc_1.26.0             rlang_0.2.2                
## [17] lazyeval_0.2.1              curl_3.2                   
## [19] S4Vectors_0.18.3            Matrix_1.2-14              
## [21] rmarkdown_1.10              pkgdown_1.1.0.9000         
## [23] labeling_0.3                desc_1.2.0                 
## [25] BiocParallel_1.14.2         readr_1.1.1                
## [27] stringr_1.3.1               munsell_0.5.0              
## [29] RCurl_1.95-4.11             DelayedArray_0.6.5         
## [31] compiler_3.5.1              xfun_0.3                   
## [33] pkgconfig_2.0.2             BiocGenerics_0.26.0        
## [35] htmltools_0.3.6             tidyselect_0.2.4           
## [37] SummarizedExperiment_1.10.1 tibble_1.4.2               
## [39] GenomeInfoDbData_1.1.0      roxygen2_6.1.0             
## [41] bookdown_0.7                IRanges_2.14.11            
## [43] matrixStats_0.54.0          fansi_0.3.0                
## [45] withr_2.1.2                 crayon_1.3.4               
## [47] dplyr_0.7.6                 MASS_7.3-50                
## [49] bitops_1.0-6                commonmark_1.5             
## [51] rappdirs_0.3.1              grid_3.5.1                 
## [53] jsonlite_1.5                gtable_0.2.0               
## [55] scales_1.0.0                cli_1.0.0                  
## [57] stringi_1.2.4               XVector_0.20.0             
## [59] fs_1.2.6                    bindrcpp_0.2.2             
## [61] xml2_1.2.0                  tools_3.5.1                
## [63] Biobase_2.40.0              glue_1.3.0                 
## [65] purrr_0.2.5                 hms_0.4.2                  
## [67] parallel_3.5.1              yaml_2.2.0                 
## [69] colorspace_1.3-2            GenomicRanges_1.32.6       
## [71] memoise_1.1.0               bindr_0.1.1</code></pre>
</div>
<div id="developer-notes" class="section level1">
<h1 class="hasAnchor">
<a href="#developer-notes" class="anchor"></a>Developer notes</h1>
<ul>
<li>The <code>S3</code> object-oriented programming paradigm is used.</li>
<li>We have adopted a functional programming style with functions and methods that often take an “object” as the first argument. This style lends itself to pipeline-style programming.</li>
<li>The GenomicDataCommons package uses the <a href="https://docs.gdc.cancer.gov/API/Users_Guide/Search_and_Retrieval/#alternative-request-format">alternative request format (POST)</a> to allow very large request bodies.</li>
</ul>
</div>
<div class="footnotes">
<hr>
<ol>
<li id="fn1"><p>See individual function and methods documentation for specific details.<a href="#fnref1" class="footnote-back">↩</a></p></li>
</ol>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#what-is-the-gdc">What is the GDC?</a></li>
      <li>
<a href="#quickstart">Quickstart</a><ul class="nav nav-pills nav-stacked">
<li><a href="#installation">Installation</a></li>
      <li><a href="#check-connectivity-and-status">Check connectivity and status</a></li>
      <li><a href="#find-data">Find data</a></li>
      <li><a href="#download-data">Download data</a></li>
      <li><a href="#metadata-queries">Metadata queries</a></li>
      </ul>
</li>
      <li><a href="#basic-design">Basic design</a></li>
      <li>
<a href="#usage">Usage</a><ul class="nav nav-pills nav-stacked">
<li><a href="#querying-metadata">Querying metadata</a></li>
      <li><a href="#authentication">Authentication</a></li>
      <li><a href="#datafile-access-and-download">Datafile access and download</a></li>
      </ul>
</li>
      <li>
<a href="#use-cases">Use Cases</a><ul class="nav nav-pills nav-stacked">
<li><a href="#cases">Cases</a></li>
      <li><a href="#files">Files</a></li>
      <li><a href="#slicing">Slicing</a></li>
      </ul>
</li>
      <li>
<a href="#troubleshooting">Troubleshooting</a><ul class="nav nav-pills nav-stacked">
<li><a href="#ssl-connection-errors">SSL connection errors</a></li>
      </ul>
</li>
      <li><a href="#sessioninfo">sessionInfo()</a></li>
      <li><a href="#developer-notes">Developer notes</a></li>
      </ul>
</div>
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by Martin Morgan, Davis Sean.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://pkgdown.r-lib.org/">pkgdown</a>.</p>
</div>

      </footer>
</div>

  

  </body>
</html>
